<a href="https://colab.research.google.com/github/mayujie/MSC_text_mining/blob/master/RD_data_preprocess.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [Amazon product data](http://jmcauley.ucsd.edu/data/amazon/index_2014.html)
# [Amazon product data Julian McAuley, UCSD](http://jmcauley.ucsd.edu/data/amazon/)
Files

"Small" subsets for experimentation


In [1]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [3]:
root_path = '/content/drive/My Drive/Colab/MSC_RD'
root_path

'/content/drive/My Drive/Colab/MSC_RD'

## **Read from text file**

In [4]:
### download text data
!gdown https://drive.google.com/uc?id=1PtUMY6KuShzR2lk_85VbNizGT_wciBNl

Downloading...
From: https://drive.google.com/uc?id=1PtUMY6KuShzR2lk_85VbNizGT_wciBNl
To: /content/reviews_text.zip
117MB [00:01, 92.1MB/s]


In [5]:
!unzip reviews_text.zip

Archive:  reviews_text.zip
  inflating: reviews_text.txt        


In [6]:
!rm /content/reviews_text.zip

In [7]:
!ls -lh *.txt

-rw-r--r-- 1 root root 303M Nov 13 11:58 reviews_text.txt


In [20]:
handle_list = []
with open('/content/reviews_text.txt', 'r') as file:
  for line in file:
    handle_list.append(line)

In [21]:
## each element in list has \n at the end
print(len(handle_list))
handle_list[1]

484826


'I highly recommend this series. It is a must for anyone who is yearning to watch "grown up" television. Complex characters and plots to keep one totally involved. Thank you Amazin Prime.\n'

## remove newline '\n' from each row

In [22]:
text_ls = map(lambda s: s.strip(), handle_list)
text_ls = list(text_ls)
text_ls[:5]

["I had big expectations because I love English TV, in particular Investigative and detective stuff but this guy is really boring. It didn't appeal to me at all.",
 'I highly recommend this series. It is a must for anyone who is yearning to watch "grown up" television. Complex characters and plots to keep one totally involved. Thank you Amazin Prime.',
 "This one is a real snoozer. Don't believe anything you read or hear, it's awful. I had no idea what the title means. Neither will you.",
 'Mysteries are interesting.  The tension between Robson and the tall blond is good but not always believable.  She often seemed uncomfortable.',
 'This show always is excellent, as far as british crime or mystery showsgoes this is one of the best ever made.  The stories are well done and the acting is top notch with interesting twists in the realistic and brutal storylines.  This show pulls no punches as it enters into the twisted minds of criminals and the profiler psychiatrist who helps out in a no

In [23]:
## read data into dataframe
import pandas as pd
df_review = pd.DataFrame(text_ls) 
df_review

0
0       I had big expectations because I love English ...
1       I highly recommend this series. It is a must f...
2       This one is a real snoozer. Don't believe anyt...
3       Mysteries are interesting.  The tension betwee...
4       This show always is excellent, as far as briti...
...                                                   ...
484821  Super bright! Awesomely Comfortable. Easy to u...
484822  I got this headlamp for my daughter since we r...
484823  This is a real bright headlamp.  I bought it f...
484824  Easy to wear headlamp that's bright. Good for ...
484825  Very nice head lamp! Bright led. Fits head wit...

[484826 rows x 1 columns]

In [24]:
del handle_list

# Preprocess data

https://towardsdatascience.com/a-beginners-guide-to-preprocessing-text-data-using-nlp-tools-5cb52a8d3cd

https://studymachinelearning.com/text-preprocessing-removal-of-punctuations/

https://gdcoder.com/all-you-need-to-know-about-nlp-text-preprocessing/

https://towardsdatascience.com/nlp-for-beginners-cleaning-preprocessing-text-data-ae8e306bef0f

https://towardsdatascience.com/nlp-building-text-cleanup-and-preprocessing-pipeline-eba4095245a0

In [25]:
%whos

Variable     Type             Data/Info
---------------------------------------
clean_text   list             n=484826
df_review    DataFrame                                 <...>[484826 rows x 1 columns]
drive        module           <module 'google.colab.dri<...>s/google/colab/drive.py'>
file         TextIOWrapper    <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
line         str              Very nice head lamp! Brig<...>and flashing modes too.\n
pd           module           <module 'pandas' from '/u<...>ages/pandas/__init__.py'>
re           module           <module 're' from '/usr/lib/python3.6/re.py'>
root_path    str              /content/drive/My Drive/Colab/MSC_RD
text_ls      list             n=484826


## 0.lowercase all letters

In [26]:
## 1. set all character to lowercase
clean_text = [sent.lower() for sent in text_ls]

In [27]:
# clean_text[:2]
clean_text[2167]

"since i didn't get a chance to watch this when it first came out, i'm glad i finally got to watch it, though i will admit that i skipped around and concentrated on episodes that also included robin or batgirl or had plots that i was more interested in.http://batmantheanimatedseries.wikia.com/wiki/robinandhttp://batmantheanimatedseries.wikia.com/wiki/batgirlinclude lists of episodes that include robin and batgirl.i figured out other interesting episodes by reading the episode summaries on amazon prime."

### replace <> punctuation as ' '


In [28]:
import re

In [29]:
clean_text = [re.sub(r'<|>', ' ', sent) for sent in clean_text]

## 1.replace 1st times URL as token \<URL\>

In [30]:
def find_noise(reg, datalist):
  count = 0
  found_list = []

  for i, row in enumerate(datalist):
    found = re.findall(reg, row)

    if found != []:
      count += 1
      # print(found[0])
      # found = list(filter(None, found[0]))
      # print(found)
      found_list.append([i ,found])
  
  return count, found_list
  # return found_list

In [31]:
## pass url 
count_url, found_url = find_noise(r"http.?://[^\s]+[\s]?|www\.[^\s|\.|\"].?[^\s\"\>\)]+|youtube\.com[^\s\)]+", clean_text)
len(found_url)

542

In [32]:
# len(found_url) ## 1752, 177, 542
found_url

[[2167,
  ['http://batmantheanimatedseries.wikia.com/wiki/robinandhttp://batmantheanimatedseries.wikia.com/wiki/batgirlinclude ']],
 [4065, ['http://www.amazon.com/burning-plain/dp/b002n4qcte/ref=pd_ys_ir__6']],
 [6236,
  ['http://www.forbes.com/sites/suwcharmananderson/2012/08/28/fake-reviews-amazons-rotten-core/http://www.wikihow.com/spot-a-fake-review-on-amazon.comhttp://www.theguardian.com/world/2013/sep/23/new-york-fake-online-reviews-yoghurt']],
 [7606,
  ['http://www.amazon.com/gp/product/b004ekmiiy/ref=cm_cr_rev_prod_title']],
 [11452,
  ['http://www.amazon.com/gp/product/b006i0gszm/ref=cm_cr_rev_prod_title']],
 [25479, ['http://www.snopes.com/photos/supernatural/mermaids.asp']],
 [33103, ['http://youtu.be/fyn1vvjcefe?t=1m40s). ']],
 [36575, ['youtube.com/watch?v=uoqmqy9akj4']],
 [37428,
  ['http://www.amazon.com/gp/product/b000xtmgtg/ref=oh_details_o01_s00_i00']],
 [38865,
  ['http://www.amazon.com/waxman-4215495n-2-inch-bumper-square/dp/b001wak6e2/ref=pd_sim_indust_3 ']],
 [3

In [33]:
## found url len > 1
# cha = []
# for i in found_url:
#   if len(i[1]) > 1:
#     # print(len(i[1]))
#     cha.append(i[0])

In [34]:
## show found url len > 1
# for a in found_url:
#   for b in cha:
#     if b == a[0]:
#       print(len(a[1]),a)

### categorize the URL (6 classes)

1.  **|facebook| --> \<FACEBOOKURL\>,**  
2.  **|book| --> \<BOOKURL\>,**
3.  **|youtube  youtu  youtu.be| --> \<YOUTUBEURL\>,**
4.  **|google  goo\.gl  goo\.| --> \<GOOGLEURL\>,**
5.  **|amzn  amazon| --> \<AMAZONURL\>,**
6.  **|rest of URL| --> \<URL\>**

In [35]:
## find keywords in found_url
def found_key_url(reg, datalist):
  found_key = []
  for i in datalist:
    for el in i[1]:
      # print(re.findall(r'amazon', el))
      # if re.findall(r'amazon|amzn', el): # |amzn  amazon|  |facebook|  |book|   |youtube  youtu  youtu.be|   |google  goo\.gl  goo\.|
      if re.findall(reg, el):
        found_key.append(i)
        # print(i)
  return found_key

In [36]:
url_facebook = found_key_url(r'facebook' ,found_url) ## 34, 0
print(len(url_facebook))
url_facebook[:10]

0


[]

In [37]:
url_book = found_key_url(r'book' ,found_url) ## 424, 0
print(len(url_book))
url_book[:10]

0


[]

In [38]:
url_amazon = found_key_url(r'amzn|amazon' ,found_url) ## 1027, 183, 501
print(len(url_amazon))
url_amazon[:10]

501


[[4065, ['http://www.amazon.com/burning-plain/dp/b002n4qcte/ref=pd_ys_ir__6']],
 [6236,
  ['http://www.forbes.com/sites/suwcharmananderson/2012/08/28/fake-reviews-amazons-rotten-core/http://www.wikihow.com/spot-a-fake-review-on-amazon.comhttp://www.theguardian.com/world/2013/sep/23/new-york-fake-online-reviews-yoghurt']],
 [7606,
  ['http://www.amazon.com/gp/product/b004ekmiiy/ref=cm_cr_rev_prod_title']],
 [11452,
  ['http://www.amazon.com/gp/product/b006i0gszm/ref=cm_cr_rev_prod_title']],
 [37428,
  ['http://www.amazon.com/gp/product/b000xtmgtg/ref=oh_details_o01_s00_i00']],
 [38865,
  ['http://www.amazon.com/waxman-4215495n-2-inch-bumper-square/dp/b001wak6e2/ref=pd_sim_indust_3 ']],
 [39465,
  ['http://www.amazon.com/gp/product/b002xl2iea/ref=oh_details_o00_s01_i01) ']],
 [41461,
  ['http://www.amazon.com/valterra-f02-4100-45%c2%b0-rv-hydroflush/product-reviews/b0002uhvaa/ref=cm_cr_dp_synop?ie=utf8&showviewpoints;=0&sortby;=bysubmissiondatedescending#r2cuau2b8xc98qit ']],
 [44374,
  

In [39]:
url_youtube = found_key_url(r'youtu|youtu.be|youtube' ,found_url) ## 80, 8, 23
print(len(url_youtube))
url_youtube[:10]

23


[[33103, ['http://youtu.be/fyn1vvjcefe?t=1m40s). ']],
 [36575, ['youtube.com/watch?v=uoqmqy9akj4']],
 [52688, ['http://youtu.be/zmqy-tmxuzu ']],
 [56402, ['youtube.com/watch?v=eco3cwrcozq']],
 [110415, ['youtube.com.this']],
 [138870, ['http://www.youtube.com/watch?v=oxz-vn6vazw[...]']],
 [148287, ['http://www.youtube.com/watch?v=oxz-vn6vazw']],
 [197730, ['https://www.youtube.com/watch?v=wucog5e5ede']],
 [229295, ['http://www.youtube.com/watch?v=cv5rwxyw8ya ']],
 [368008,
  ['youtube.com\\im14pinballthis',
   'youtube.com\\im14pinballor',
   'youtube.com/im14pinballthanks!!']]]

In [40]:
url_google = found_key_url(r'google|goo\.gl|goo\.' ,found_url) ## 80, 0, 13
print(len(url_google))
url_google[:10]

13


[[376778, ['http://goo.gl/cjnwsh ']],
 [389752, ['http://goo.gl/3tly4x ']],
 [389828, ['http://goo.gl/3tly4x ']],
 [396016, ['http://goo.gl/0igojv ']],
 [472083, ['http://goo.gl/iz0vgw ']],
 [472260, ['http://goo.gl/iz0vgw ']],
 [472337, ['http://goo.gl/iz0vgw ']],
 [474088, ['http://goo.gl/iz0vgw ']],
 [476743, ['http://goo.gl/iz0vgw ']],
 [476925, ['http://goo.gl/iz0vgw ']]]

In [42]:
clean_text[148287]

"this is essential if you plan to use your pen regularly. it will save you money. buy some good ink (read the reviews) and if you travel, you should also have a set of cartridges for backup). very simple piece of plastic that does the job. look at the different sellers to get the best deal but this piece is worth the small price. finally, there are two schools of thought on filling this. they both have the same outcome; ink is inside the converter. lamy recommends filling it while it is attached to the pen and filling it directly through the nib. if you have a deep, empty bottle this is for sure the best route. the video from lamy follows; he covers the z26 and z24.. but it's the same procedure. http://www.youtube.com/watch?v=oxz-vn6vazw"

### replace
1.  **|facebook| --> \<FACEBOOKURL\>,**  
2.  **|book| --> \<BOOKURL\>,**
3.  **|youtube  youtu  youtu.be| --> \<YOUTUBEURL\>,**
4.  **|google  goo\.gl  goo\.| --> \<GOOGLEURL\>,**
5.  **|amzn  amazon| --> \<AMAZONURL\>,**
6.  **|rest of URL| --> \<URL\>**

In [43]:
## replace to <FACEBOOK_URL>
for i in url_facebook:
  clean_text[i[0]] = re.sub(r"http.?://[^\s]+[\s]?|www\.[^\s|\.|\"].?[^\s\"\>\)]+|youtube\.com[^\s\)]+", " <FACEBOOKURL> ", clean_text[i[0]])

In [44]:
## replace to <BOOK_URL>
for i in url_book:
  clean_text[i[0]] = re.sub(r"http.?://[^\s]+[\s]?|www\.[^\s|\.|\"].?[^\s\"\>\)]+|youtube\.com[^\s\)]+", " <BOOKURL> ", clean_text[i[0]])

In [45]:
## replace to <YOUTUBE_URL>
for i in url_youtube:
  # print(clean_text[i[0]])
  clean_text[i[0]] = re.sub(r"http.?://[^\s]+[\s]?|www\.[^\s|\.|\"].?[^\s\"\>\)]+|youtube\.com[^\s\)]+", " <YOUTUBEURL> ", clean_text[i[0]])

In [46]:
## replace to <GOOGLE_URL>
for i in url_google:
  # print(clean_text[i[0]])
  clean_text[i[0]] = re.sub(r"http.?://[^\s]+[\s]?|www\.[^\s|\.|\"].?[^\s\"\>\)]+|youtube\.com[^\s\)]+", " <GOOGLEURL> ", clean_text[i[0]])

In [47]:
## replace to <AMAZON_URL>
for i in url_amazon:
  clean_text[i[0]] = re.sub(r"http.?://[^\s]+[\s]?|www\.[^\s|\.|\"].?[^\s\"\>\)]+|youtube\.com[^\s\)]+", " <AMAZONURL> ", clean_text[i[0]])

In [48]:
## 2. replace url as token
clean_text = [re.sub(r"http.?://[^\s]+[\s]?|www\.[^\s|\.|\"].?[^\s\"\>\)]+|youtube\.com[^\s\)]+", " <URL> ", sent) for sent in clean_text]

In [49]:
clean_text[2167]

"since i didn't get a chance to watch this when it first came out, i'm glad i finally got to watch it, though i will admit that i skipped around and concentrated on episodes that also included robin or batgirl or had plots that i was more interested in. <URL> lists of episodes that include robin and batgirl.i figured out other interesting episodes by reading the episode summaries on amazon prime."

In [50]:
clean_text[138870]

"obviously this pen is not a very expensive fountain pen that is a 'showpiece', rather it is an everyday pen that should survive daily use and is a replacement for a ballpoint or rollerball pen. it takes a bit of use to get formalized to using a fountain pen and sometimes you may find it skipping especially when it has air pockets in the chamber or you're at a strange angle. this is expected but that is the fun and uniqueness of using such a pen. it is plastic but very sold (not weight wise but feel wise). the clip is large but this is a good thing if you carry it in your pocket and want it to be secure. the nib is a fine nib but from reading other sites, it seems as if the size is closer to a medium. (japan nibs vs. german?). whatever; the script is very close to my other nibs labeled as medium. this is a very good everyday thickness; not too fine and not too broad. the one cartridge it comes with is blue. i would recommend buying a good bottle of ink and a z24 converter. this up fron

## 1.replace 2nd time URL as token

In [51]:
def find_noise_url(reg, datalist):
  # count = 0
  found_list = []

  for i, row in enumerate(datalist):
    found = re.findall(reg, row)

    if found != []:
      # count += 1
      # print(found[0])
      found = list(filter(None, found[0]))
      # print(found)
      found_list.append([i ,found])
  
  # return count, found_list
  return found_list

In [52]:
## pass url level 2
# count_url, found_url = find_noise(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", clean_text)
found_url = find_noise_url(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", clean_text)
len(found_url)

158

In [53]:
## found url level 2 len > 1
cha = []
for i in found_url:
  if len(i[1]) > 1:
    # print(len(i[1]))
    cha.append(i[0])

In [54]:
## show found url level 2 len > 1
for a in found_url:
  for b in cha:
    if b == a[0]:
      print(len(a[1]), a)

In [55]:
clean_text[158069]

'i was really surprised at not only the quality, but also the thought that went into these view-tab dividers.  these aren&#8217;t your tabs that you buy in your local big box store, but rather tabs that will last a long time.  they are made of a thick poly and you can change out your project many times over and they&#8217;ll hold up.instead of marking on and inserting a small piece of paper into open tabs, these are solid see through.  what you do is set up your titles, print them out, and insert the sheet into a clear view pocket.  the durability factor is a real plus and the idea rather ingenious and eco-friendly.  you won&#8217;t have to purchase new tabs or divider sheets each time you have a new project.if you are uncertain as to how to format your pages so they&#8217;ll show up on the side, there are full instructions in an insert.  this instructional should be tucked away somewhere in your notebook or ring binder so you can use it again.  instructions are laid out for corel word

In [56]:
clean_text[177711]

'canon pixma mg5420 wireless color photo printer with scanner and copier[12/13/2013 note: i suggest also looking at the canon pixma mg7120 if you are interested in the mg5420][12/22/2013 note: i added a section below noting why the mg5420 is superior to the mg5520 in my opinion]overview:----------if you don\'t care about having gray or ethernet, but want great photo quality in an inexpensive all-in-one style (copier, scanner) ink jet printer, this can be a quality/value sweet spot for you in the canon pixma line giving you a lot of bang for the buck.  this is the replacement for mg5320, only available since december 2012.if this seems right for you, consider looking at the slightly more upscale canon mg7120 features before committing as the cost difference isn\'t that much.pros:------- a well regarded consumer magazine considers these mg5000/mg6000/mg8000 series printers the best photo quality for all-in-one (copier, scanner) ink jets.  this model is the entry point into this line.- th

In [73]:
## find keywords in found_url level 2
for i in found_url:
  for el in i[1]:
    # print(re.findall(r'amazon', el))
    if re.findall(r'amazon', el): # amzn  amazon  facebook  book   youtube  youtu  youtu.be   google  goo\.gl  goo\. 
      print(i)

[151993, ['amazon.com/craigs']]
[153495, ['amazon.com/amazonbasics']]
[317903, ['amazon.com/vine']]
[440511, ['redamazon.com/maglite-s6d035-display-heavy-duty-flashlight/dp/b0002jt3k2/ref=sr_1_1?ie=utf8&qid;=1375235931&sr;=8-1&keywords;=maglite+s6d035']]


In [62]:
# len(found_url) ## 2246 # 96 # 158
found_url ## 575 # 96 # 158

[[8325, ['hate...love/hate']],
 [40992, ['buy.edit/update']],
 [50916, ['weight.fit/finish']],
 [52100, ['actron.com/downloads/fifth']],
 [57341, ['automatic.com/cars']],
 [57342, ['automatic.com/apps']],
 [67392, ['rap...rock/metal']],
 [68264, ['such.ac/dc']],
 [69621, ['grand....rock/pop']],
 [72850, ['t.n.t.ac/dc']],
 [75611, ['list.hits/chart']],
 [77644, ['blazemonger.com/gg']],
 [94599, ['j.lo/richard']],
 [99177, ['others.rap/rock']],
 [101084, ['55.in/out']],
 [102449, ['listen.to/wwsu']],
 [105961, ['dr.dre/ice']],
 [107463, ['alt.rock/indie']],
 [108234, ["v.good/average.i'm"]],
 [108378, ['maverick.com/alanis-']],
 [109364, ['bricksmyspace.com/getthebloodmoney']],
 [109547, ['gatesmyspace.com/getthebloodmoney']],
 [109803, ['itmyspace.com/getthebloodmoney']],
 [110039, ['upmyspace.com/getthebloodmoney']],
 [113107, ['possible.band/guests:frank']],
 [114434, ['myspace.com/yingyangtwins']],
 [114559, ['myspace.com/theshowno']],
 [114734, ['myspace.com/theshowno']],
 [114739, 

### categorize the URL 2nd time (6 classes)

1.  **|facebook| --> \<FACEBOOKURL\>,**  
2.  **|book| --> \<BOOKURL\>,**
3.  **|youtube  youtu  youtu.be| --> \<YOUTUBEURL\>,**
4.  **|google  goo\.gl  goo\.| --> \<GOOGLEURL\>,**
5.  **|amzn  amazon| --> \<AMAZONURL\>,**
6.  **|rest of URL| --> \<URL\>**

In [74]:
url_facebook = found_key_url(r'facebook' ,found_url) ## 34
print(len(url_facebook))
url_facebook[:10]

2


[[293352, ['facebook.com/greenmountaincoffee']],
 [300249, ['facebook.com/miraclenoodle']]]

In [75]:
# clean_text[268281]

In [76]:
url_book = found_key_url(r'book' ,found_url) ## 424
print(len(url_book))
url_book[:10]

3


[[175731, ['scanners.book/object']],
 [293352, ['facebook.com/greenmountaincoffee']],
 [300249, ['facebook.com/miraclenoodle']]]

In [77]:
# clean_text[1146694]
clean_text[175731]

'i\'ve had theepson b11b198011 perfection v600 photo scannerfor two years. i compared it side-by-side with the v370 using the epson software and also vuescan by hamrick software.here is the concise comparison, based on the most common scanning tasks:quick document scan: both scanners handle this with one button. both use leds instead of fluorescent bulbs, so there is no warm-up delay. there\'s virtually no difference between them. use the included epson software.occasional photo scans: both scanners excel at this, again with one button. no difference between them. use the epson softwareocr: both scanners include ocr software. depending on your application, both do an acceptable job. take your time, line up the pages carefully, rescan when a page has lots of errors. neither scanner has an automatic document feeder, so you won\'t be hand-feeding 100 pages without some fatigue. no difference between the scanners.book/object scans: this is something i didn\'t know i would need before getti

In [78]:
url_youtube = found_key_url(r'youtu|youtu.be|youtube' ,found_url) ## 80
print(len(url_youtube))
url_youtube[:10]

0


[]

In [79]:
clean_text[178325]

'the idea behind the ampad is you can take notes on these and use their free iphone app to transfer these into an electronic storable form, like evernote or dropbox.  the app has you line up the special boxes in the four corners of each page and take a photo.  the software then processes the image to make it more readable, and to eliminate any non-paper from the final image.  you can then upload to your chosen cloud service.  it works as advertised.  my only major complaint is they do not allow you to bundle together multiple images into a single document.however, i would not recommend most users to buy this.when it comes down to it, you are paying $5+ for each pad which contains 40 sheets.  a comparable writing pad from wal-mart or target can be had for $2 (or 25 cents if you buy during back-to-school sales).  the same results can be had with these cheaper pads by purchasing an app like jotnot pro (99 cents) (...)itunes.apple.com/us/app/jotnot-scanner-pro-scan-multipage/id307868751?mt

In [80]:
url_google = found_key_url(r'google|goo\.gl|goo\.' ,found_url) ## 80
print(len(url_google))
url_google[:10]

0


[]

In [81]:
# clean_text[582422]

In [82]:
url_amazon = found_key_url(r'amzn|amazon' ,found_url) ## 1027
print(len(url_amazon))
url_amazon[:10]

4


[[151993, ['amazon.com/craigs']],
 [153495, ['amazon.com/amazonbasics']],
 [317903, ['amazon.com/vine']],
 [440511,
  ['redamazon.com/maglite-s6d035-display-heavy-duty-flashlight/dp/b0002jt3k2/ref=sr_1_1?ie=utf8&qid;=1375235931&sr;=8-1&keywords;=maglite+s6d035']]]

In [83]:
# clean_text[950979]

### replace
1.  **|facebook| --> \<FACEBOOKURL\>,**  
2.  **|book| --> \<BOOKURL\>,**
3.  **|youtube  youtu  youtu.be| --> \<YOUTUBEURL\>,**
4.  **|google  goo\.gl  goo\.| --> \<GOOGLEURL\>,**
5.  **|amzn  amazon| --> \<AMAZONURL\>,**
6.  **|rest of URL| --> \<URL\>**

In [84]:
## replace to <FACEBOOKURL>
for i in url_facebook:
  clean_text[i[0]] = re.sub(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", " <FACEBOOKURL> ", clean_text[i[0]])

## replace to <BOOKURL>
for i in url_book:
  clean_text[i[0]] = re.sub(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", " <BOOKURL> ", clean_text[i[0]])

## replace to <YOUTUBEURL>
for i in url_youtube:
  # print(clean_text[i[0]])
  clean_text[i[0]] = re.sub(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", " <YOUTUBEURL> ", clean_text[i[0]])

## replace to <GOOGLEURL>
for i in url_google:
  # print(clean_text[i[0]])
  clean_text[i[0]] = re.sub(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", " <GOOGLEURL> ", clean_text[i[0]])

## replace to <AMAZONURL>
for i in url_amazon:
  clean_text[i[0]] = re.sub(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", " <AMAZONURL> ", clean_text[i[0]])

In [85]:
## 3. replace 2nd level url as token
clean_text = [re.sub(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", " <URL> ", sent) for sent in clean_text]

In [86]:
# clean_text[963498]

## 2.replace email and @mention as token \<MENTION\>

In [87]:
## replace email as token
clean_text = [re.sub(r'@\w+|[\w\.-]+@[\w\.-]+\.\w+|[\w\.-]+@[\w\.-]+', ' <MENTION> ', sent) for sent in clean_text]

In [88]:
clean_text[80104]

"yo this is like, awesome...it's closer to all eyez on me than most people make it out to be...pac's vision on this album is so inspiring and his delivery is much better than on his earlier releases...r.i.p. 2pac"

In [89]:
clean_text[10323]

'i just finished watching "dead hooker in a trunk" by and starring the soska twins.  jen and sylvia play the leads in one seriously messed up grindhouse-y indie horror film (their first).  originally planned to star four female leads, someone apparently dropped out and one of the characters was re-written days before shooting began to be a male.so while this would have been a different film if it starred and entirely female cast, i have to say the inclusion of c.j. wallis as "goody two shoes" was a stroke of luck.  although he was busy co-starring in the film, as noted on imdb he also "...edited, shot, sound tracked, marketed, and did all post on the film."  he is absolutely hilarious in this film and show a key sense of comedic timing.as for the soska twins...what can i say.  smart, sexy, dark and a little disturbing...well a lot disturbing.  at one point, i realized that they might be just playing themselves (seemed so natural as loopy twins).so how was the movie?  for an indie first

## 3.replace $28.. as token \<PRICE\>

In [90]:
## replace dollor sign as price token
clean_text = [re.sub(r'\$\d+(?:\,\d+|\.\d+)?', ' <PRICE> ', sent) for sent in clean_text]

In [91]:
clean_text[1213]

'this was free on amazon prime. after watching for three and one half seasons, suddenly they want  <PRICE>  plus meaning about  <PRICE> . feel like this is a bait and switch. amazon prime offered me  <PRICE>  and told me to complain to the studio. feels like reading half a book and losing it.'

In [92]:
clean_text[1423]

'the question is "did we really need to rebuild her?" after watch all eight episodes of this bionic woman remake, i say no!this last tv season nbc choose to remake the bionic woman for the 2000\'s. in trying to gain success like sci-fi channel had the other 1970\'s remake  battlestar galactica, the producers took a gritter style...and it failed on many levels.as lovely as michelle ryan is, she is no lindsay wagner. ms ryan tries to give this piece heart. the character , a bartender , is given her bionics after an auto accident with her government surgeon boyfriend.  she is saddled her with a felony kid sister (lucy hale) who was dumped on her when their father disappeared. ryan is one of the reason\'s to watch this show.however, the heart of the show is buried deep within sub plots of the previous sort of evilish bionic woman sarah (katee sackhoff from the remake of gallatica) and ideas of double crossing people that seemed to be ripped off from old alias plots. this new woman is told 

## 4.replace time with space

In [93]:
## replace time with space ' '
clean_text = [re.sub(r'\d+:\d+:\d+|\d+\-\d+\-\d+', ' ', sent) for sent in clean_text]

## 5.replace unicode decimal code

https://www.w3schools.com/python/python_regex.asp

https://dev.w3.org/html5/html-author/charref

https://stackoverflow.com/questions/2087370/decode-html-entities-in-python-string

https://wiki.python.org/moin/EscapingHtml

https://docs.python.org/3/library/html.html?highlight=escape#html.escape

https://www.htmlhelp.com/reference/html40/entities/special.html

In [94]:
# text_ls[261620]

In [95]:
# clean_text[261620]

In [96]:
## return all the html entities to symbol
# clean_text = [html.unescape(i)  for i in clean_text]

In [97]:
count_symbol, symbol = find_noise(r'&\w+;', clean_text) ## &quot; &lt;|&gt;|&le;|&ge;   &\w+;
len(symbol) # 1st round 60078  # 2nd round  5773 #### final version 14205， 16408

16408

In [98]:
print(clean_text[75390])
symbol[:15]

while its not the most essential madonna album, bedtime stories is a nice breath of fresh air. following the overhyped and dissapointing erotica, bedtime stories is set on a new pair of sheets and a mint on the pilow. the album contains a handful of sweet dreams; like the smooth and sultry "secret" and "take a bow" is probably the best ballad since 1993's "rain". madonna has always been a dance artist though, and on this particular album she supplies some of the electronic beats that will lead to "ray of light". "holiday" was really where it all started, and the bjork/nellee hooper produced "bedtime story" shows madonna has not become a contemporary artist. for those snots who have snubbed their noses at madonna for all the sex hoopla, she gives them the finger on "human nature" with an infectious r&b groove and a potty mouth. "oops.. i didnt know i couldnt talk about sex.." she sarcastically spits. meooww! check the video if you think shes softened at all. as for the duds, theres good

[[298, ['&lt;']],
 [311, ['&eacute;']],
 [516, ['&a;']],
 [595, ['&lt;', '&lt;']],
 [1110, ['&eacute;']],
 [1116, ['&eacute;']],
 [1149, ['&eacute;']],
 [1184, ['&eacute;']],
 [1197, ['&eacute;']],
 [1272, ['&o;', '&e;', '&order;']],
 [1609, ['&b;']],
 [1715, ['&eacute;']],
 [1840, ['&eacute;']],
 [1842, ['&eacute;']],
 [1879, ['&eacute;']]]

In [ ]:
# clean_text[770708]

In [99]:
ss = []
for i in symbol:
  for j in i[1]:
    ss.append(j)

In [100]:
from collections import Counter
Counter(ss).most_common()  # &amp; &  &apos; '  &lt; <  &gt; >

[('&quot;', 111776),
 ('&amp;', 4708),
 ('&b;', 2601),
 ('&eacute;', 1327),
 ('&lt;', 801),
 ('&d;', 692),
 ('&w;', 358),
 ('&ntilde;', 256),
 ('&m;', 220),
 ('&egrave;', 205),
 ('&t;', 199),
 ('&s;', 153),
 ('&aacute;', 139),
 ('&n;', 130),
 ('&oacute;', 115),
 ('&iexcl;', 112),
 ('&ouml;', 100),
 ('&iacute;', 88),
 ('&g;', 86),
 ('&c;', 80),
 ('&gt;', 72),
 ('&e;', 70),
 ('&uuml;', 60),
 ('&j;', 50),
 ('&h;', 47),
 ('&ms;', 47),
 ('&mdash;', 46),
 ('&auml;', 44),
 ('&agrave;', 41),
 ('&p;', 37),
 ('&a;', 32),
 ('&r;', 32),
 ('&o;', 28),
 ('&decker;', 25),
 ('&f;', 24),
 ('&szlig;', 22),
 ('&uacute;', 20),
 ('&l;', 19),
 ('&bull;', 14),
 ('&iquest;', 9),
 ('&y;', 8),
 ('&scan;', 8),
 ('&cheese;', 8),
 ('&k;', 7),
 ('&white;', 6),
 ('&save;', 6),
 ('&the;', 5),
 ('&fnr;', 5),
 ('&honey;', 5),
 ('&ws;', 4),
 ('&delivered;', 4),
 ('&tg;', 4),
 ('&i;', 4),
 ('&half;', 4),
 ('&lock;', 4),
 ('&seymour;', 4),
 ('&country;', 3),
 ('&mjg;', 3),
 ('&shout;', 3),
 ('&fee;', 3),
 ('&heavy;', 3),


In [101]:
# for i in Counter(ss).most_common():
#   print(i[0], html.unescape(i[0]), i[1])

In [102]:
!pip install namedentities

In [103]:
from namedentities import *
entities('&trade; &amp; &lt; &apos; &gt; &rsquo; &lsquo;', 'decimal')

'&#8482; &amp; &lt; &apos; &gt; &#8217; &#8216;'

In [104]:
# for i in symbol:
#   print(list(map(lambda x: entities(x, 'decimal'), i[1])))

In [105]:
## replace Html Entity to decimal code
clean_text = [entities(i, 'decimal') for i in clean_text]

In [106]:
## remove &lt; <  &gt; >
clean_text = [re.sub(r'&lt;|&gt;', ' ', sent) for sent in clean_text]

In [107]:
## replace &amp; &  &apos; '
clean_text = [re.sub(r'&amp;', '&', sent) for sent in clean_text]
clean_text = [re.sub(r'&apos;', "'", sent) for sent in clean_text]
clean_text = [re.sub(r'&quot;', '"', sent) for sent in clean_text]

In [108]:
text_ls[51896]

'On my 60in hilift extreme, this thing is like a fat girl in spandex.  Without another reviewers trick (start the zipper a few inches, remove the base, then slide on), I would never have been able to get the infernal thing on.'

In [109]:
clean_text[51896]

'on my 60in hilift extreme, this thing is like a fat girl in spandex.  without another reviewers trick (start the zipper a few inches, remove the base, then slide on), i would never have been able to get the infernal thing on.'

In [110]:
clean_text[43148]

"the engine in my 93 pathfinder will die when first started for the day especially during cold weather. my mechanic suggested trying some type of fuel system cleaner. i try to use chevron gas because it has techron, in minimal amount of course but after all these years it may be dirty in there. i love this car so i like to keep it running well.i say it seems to have helped because it does run better now but it is also warming up. i will continue to add this every 3000 miles or so and see if i have any issues next winter. i don't drive this car that often so i won't need a lot. which also may be why it's gummed up. wish me luck!"

In [111]:
clean_text[69437]

'you say: there are a few tard muffins on this review page. lyrically she is the profoundest and most talented of all we see today and yesterday, and song craft wise. even a sarah maclachlan cannot live up to the lyricism from an educated viewpoint, although the heartfelt reachiness of sarah makes up for that by far. not to mention the charitible nature. this album has prevented suicides and provided wisdom and guidance. night visions understands the lowest points of life. i\'m not retarded when i hear luka, i\'m justified. as with the song "what\'s the matter here" by 10,000 maniacs. "neighbor hood girls" is me..it\'s off her first. fancy poultry/iron bound market is one of the best songs ever because i wrote a similar diddy as a child, (exactly copying a the intro to the latter song without knowing about susanne\'s years before..) thank you dear for saving my life, and along with "days of open hand", this cd is an instruction manual for living.'

In [112]:
## find unicode decimal
count_encode, found_encode = find_noise(r'\&\#\d+\;{1}?', clean_text) 

In [113]:
print(count_encode, len(found_encode))
## found 64701, 2275， 4498 unicode decimal code
found_encode[:10]

4498 4498


[[34,
  ['&#8217;',
   '&#8217;',
   '&#8217;',
   '&#8217;',
   '&#8217;',
   '&#8217;',
   '&#8220;',
   '&#8221;',
   '&#8220;',
   '&#8221;',
   '&#8220;',
   '&#8221;',
   '&#8220;',
   '&#8221;',
   '&#8220;',
   '&#8217;',
   '&#8221;',
   '&#8217;',
   '&#8217;',
   '&#8217;',
   '&#8217;',
   '&#8220;',
   '&#8221;']],
 [68, ['&#128522;']],
 [237, ['&#62;']],
 [311, ['&#233;']],
 [485, ['&#8217;', '&#8212;', '&#8217;']],
 [731, ['&#8230;']],
 [1110, ['&#233;']],
 [1116, ['&#233;']],
 [1149, ['&#233;']],
 [1184, ['&#233;']]]

In [114]:
encode_el = []
for j in found_encode:
  for el in j[1]:
    encode_el.append(el)

In [115]:
print(len(encode_el)) # 618526 values  404049, 12781， 20090
print(len(set(encode_el))) # 801, 108， 169 uniqe values
# set(encode_el)

20090
169


### check emoji

https://pypi.org/project/emoji/

https://www.quackit.com/character_sets/emoji/emoji_v3.0/unicode_emoji_v3.0_characters_all.cfm

https://www.w3schools.com/charsets/ref_emoji.asp


In [116]:
!pip install emoji --upgrade

     |████████████████████████████████| 51kB 2.8MB/s 
  Created wheel for emoji: filename=emoji-0.6.0-cp36-none-any.whl size=49716 sha256=f1e2131f85a32dcc8e1aeeba11192332ef3e40b482388087556a958dbc5cbebd
  Stored in directory: /root/.cache/pip/wheels/46/2c/8b/9dcf5216ca68e14e0320e283692dce8ae321cdc01e73e17796
Successfully built emoji


In [117]:
from collections import Counter
import emoji
import html

html_entity_unique = Counter(encode_el).most_common()
converted = [(i[0], i[1], html.unescape(i[0])) for i in html_entity_unique]
converted

[('&#8217;', 7821, '’'),
 ('&#8221;', 2161, '”'),
 ('&#8220;', 1793, '“'),
 ('&#233;', 1327, 'é'),
 ('&#8211;', 1095, '–'),
 ('&#8230;', 789, '…'),
 ('&#65533;', 617, '�'),
 ('&#9658;', 502, '►'),
 ('&#1150;', 487, 'Ѿ'),
 ('&#8212;', 339, '—'),
 ('&#62;', 289, '>'),
 ('&#8216;', 282, '‘'),
 ('&#241;', 256, 'ñ'),
 ('&#232;', 205, 'è'),
 ('&#8226;', 170, '•'),
 ('&#60;', 158, '<'),
 ('&#225;', 139, 'á'),
 ('&#243;', 115, 'ó'),
 ('&#161;', 112, '¡'),
 ('&#246;', 100, 'ö'),
 ('&#237;', 88, 'í'),
 ('&#8201;', 66, '\u2009'),
 ('&#252;', 60, 'ü'),
 ('&#9670;', 58, '◆'),
 ('&#9702;', 52, '◦'),
 ('&#9650;', 52, '▲'),
 ('&#9642;', 46, '▪'),
 ('&#228;', 44, 'ä'),
 ('&#224;', 41, 'à'),
 ('&#2951;', 40, 'இ'),
 ('&#9654;', 36, '▶'),
 ('&#9830;', 35, '♦'),
 ('&#10004;', 34, '✔'),
 ('&#9660;', 33, '▼'),
 ('&#9788;', 33, '☼'),
 ('&#1411;', 29, 'փ'),
 ('&#8203;', 28, '\u200b'),
 ('&#4335;', 24, 'ჯ'),
 ('&#10003;', 24, '✓'),
 ('&#9679;', 23, '●'),
 ('&#223;', 22, 'ß'),
 ('&#250;', 20, 'ú'),
 ('&#9674;', 

In [118]:
import emoji
for el in converted:
  print('&#' + str(ord(el[2])) + ';', '\t', el[1], '\t', html.unescape('&#' + str(ord(el[2])) + ';'), '\t', emoji.demojize(el[2]))
  # print(ord(el[2]))

&#8217; 	 7821 	 ’ 	 ’
&#8221; 	 2161 	 ” 	 ”
&#8220; 	 1793 	 “ 	 “
&#233; 	 1327 	 é 	 é
&#8211; 	 1095 	 – 	 –
&#8230; 	 789 	 … 	 …
&#65533; 	 617 	 � 	 �
&#9658; 	 502 	 ► 	 ►
&#1150; 	 487 	 Ѿ 	 Ѿ
&#8212; 	 339 	 — 	 —
&#62; 	 289 	 > 	 >
&#8216; 	 282 	 ‘ 	 ‘
&#241; 	 256 	 ñ 	 ñ
&#232; 	 205 	 è 	 è
&#8226; 	 170 	 • 	 •
&#60; 	 158 	 < 	 <
&#225; 	 139 	 á 	 á
&#243; 	 115 	 ó 	 ó
&#161; 	 112 	 ¡ 	 ¡
&#246; 	 100 	 ö 	 ö
&#237; 	 88 	 í 	 í
&#8201; 	 66 	   	  
&#252; 	 60 	 ü 	 ü
&#9670; 	 58 	 ◆ 	 ◆
&#9702; 	 52 	 ◦ 	 ◦
&#9650; 	 52 	 ▲ 	 ▲
&#9642; 	 46 	 ▪ 	 :black_small_square:
&#228; 	 44 	 ä 	 ä
&#224; 	 41 	 à 	 à
&#2951; 	 40 	 இ 	 இ
&#9654; 	 36 	 ▶ 	 :play_button:
&#9830; 	 35 	 ♦ 	 :diamond_suit:
&#10004; 	 34 	 ✔ 	 :heavy_check_mark:
&#9660; 	 33 	 ▼ 	 ▼
&#9788; 	 33 	 ☼ 	 ☼
&#1411; 	 29 	 փ 	 փ
&#8203; 	 28 	 ​ 	 ​
&#4335; 	 24 	 ჯ 	 ჯ
&#10003; 	 24 	 ✓ 	 ✓
&#9679; 	 23 	 ● 	 ●
&#223; 	 22 	 ß 	 ß
&#250; 	 20 	 ú 	 ú
&#9674; 	 20 	 ◊ 	 ◊
&#65292; 	 20 	 ， 	 ，
&#6

In [119]:
emoji_set = [emoji.demojize(el[2]) for el in converted]
print(len(emoji_set)) ## 801, 108, 169
emoji_set[:15]

169


['’', '”', '“', 'é', '–', '…', '�', '►', 'Ѿ', '—', '>', '‘', 'ñ', 'è', '•']

In [120]:
token_emoji = [re.findall(r'\:\w+\:', i) for i in emoji_set if re.findall(r'\:\w+\:', i)]
print(len(token_emoji)) # 157, 17, 24
token_emoji

24


[[':black_small_square:'],
 [':play_button:'],
 [':diamond_suit:'],
 [':heavy_check_mark:'],
 [':smiling_face:'],
 [':thumbs_up:'],
 [':heart_suit:'],
 [':trade_mark:'],
 [':red_heart:'],
 [':glowing_star:'],
 [':smiling_face_with_smiling_eyes:'],
 [':winking_face:'],
 [':grinning_face_with_smiling_eyes:'],
 [':beaming_face_with_smiling_eyes:'],
 [':frowning_face:'],
 [':ballot_box_with_check:'],
 [':frowning_face_with_open_mouth:'],
 [':winking_face_with_tongue:'],
 [':smiling_face_with_sunglasses:'],
 [':smiling_face_with_halo:'],
 [':squinting_face_with_tongue:'],
 [':face_savoring_food:'],
 [':confused_face:'],
 [':double_exclamation_mark:']]

In [121]:
unicode_2_emoji = {'&#' + str(ord(emoji.emojize(i[0]))) + ';' : ' '+i[0]+' ' for i in token_emoji}
print(len(unicode_2_emoji.keys())) # 157, 17, 24
unicode_2_emoji

24


{'&#10004;': ' :heavy_check_mark: ',
 '&#10084;': ' :red_heart: ',
 '&#127775;': ' :glowing_star: ',
 '&#128077;': ' :thumbs_up: ',
 '&#128513;': ' :beaming_face_with_smiling_eyes: ',
 '&#128516;': ' :grinning_face_with_smiling_eyes: ',
 '&#128519;': ' :smiling_face_with_halo: ',
 '&#128521;': ' :winking_face: ',
 '&#128522;': ' :smiling_face_with_smiling_eyes: ',
 '&#128523;': ' :face_savoring_food: ',
 '&#128526;': ' :smiling_face_with_sunglasses: ',
 '&#128533;': ' :confused_face: ',
 '&#128540;': ' :winking_face_with_tongue: ',
 '&#128541;': ' :squinting_face_with_tongue: ',
 '&#128550;': ' :frowning_face_with_open_mouth: ',
 '&#8252;': ' :double_exclamation_mark: ',
 '&#8482;': ' :trade_mark: ',
 '&#9642;': ' :black_small_square: ',
 '&#9654;': ' :play_button: ',
 '&#9745;': ' :ballot_box_with_check: ',
 '&#9785;': ' :frowning_face: ',
 '&#9786;': ' :smiling_face: ',
 '&#9829;': ' :heart_suit: ',
 '&#9830;': ' :diamond_suit: '}

### save to dictionary

In [122]:
# import json

# with open('/content/drive/My Drive/Colab/MSC_RD/unicode_2_emoji.json', 'w') as fp:
#     json.dump(unicode_2_emoji, fp)

In [123]:
import json

with open('/content/drive/My Drive/Colab/MSC_RD/unicode_2_emoji.json', 'r') as fp:
    dict_unicode_2_emoji = json.load(fp)

print(len(dict_unicode_2_emoji.keys())) ## 157
dict_unicode_2_emoji

157


{'&#10002;': ' :black_nib: ',
 '&#10004;': ' :heavy_check_mark: ',
 '&#10006;': ' :heavy_multiplication_x: ',
 '&#10017;': ' :star_of_David: ',
 '&#10024;': ' :sparkles: ',
 '&#10052;': ' :snowflake: ',
 '&#10060;': ' :cross_mark: ',
 '&#10071;': ' :exclamation_mark: ',
 '&#10083;': ' :heavy_heart_exclamation: ',
 '&#10084;': ' :red_heart: ',
 '&#11088;': ' :white_medium_star: ',
 '&#11093;': ' :heavy_large_circle: ',
 '&#127756;': ' :milky_way: ',
 '&#127774;': ' :sun_with_face: ',
 '&#127775;': ' :glowing_star: ',
 '&#127776;': ' :shooting_star: ',
 '&#127800;': ' :cherry_blossom: ',
 '&#127801;': ' :rose: ',
 '&#127802;': ' :hibiscus: ',
 '&#127803;': ' :sunflower: ',
 '&#127808;': ' :four_leaf_clover: ',
 '&#127867;': ' :clinking_beer_mugs: ',
 '&#127872;': ' :ribbon: ',
 '&#127880;': ' :balloon: ',
 '&#127881;': ' :party_popper: ',
 '&#127882;': ' :confetti_ball: ',
 '&#127909;': ' :movie_camera: ',
 '&#127932;': ' :musical_score: ',
 '&#128044;': ' :dolphin: ',
 '&#128051;': ' :s

In [124]:
for i in found_encode:
  for key in unicode_2_emoji:
    clean_text[i[0]] = clean_text[i[0]].replace(key, unicode_2_emoji[key])

In [125]:
# clean_text[1772326]

In [126]:
# clean_text[893155]

In [127]:
## replace &#34; &#8220; &#8221; sign back to Quotation Mark " 
clean_text = [re.sub(r'&#34;|&#8220;|&#8221;', '"', sent) for sent in clean_text]

In [128]:
text_ls[829]

'Yet another &#34;apocolypse&#34; show written by &#34;big city dwellers&#34; who have a total &#34;city centered&#34; view of EVERYTHING.Let me start by sayiing that the premise of the shows &#34;disaster&#34; is not too bad.  We are drawn in by being left guessing as to the nature of what has happened.This is where it wanders off the path.  The small town of Jericho is nothing more than a minature version of New York or LA from other movies.  with the exception of one or two &#34;farms,&#34; EVERYTHING ceases to exist at the town limits.  A character is either &#34;of the town&#34; or an &#34;outsider&#34; and that is it.It is plainly obvious that the writers have never spent any time in the real world, outside of a big city.  Rural America does not work like this.  Small towns blur.  The town limits rarely really limit the towns themselves and they often sort of blend into nearby communities and they really do encompass surrounding rural areas.  This is not europe and the local twon

In [129]:
clean_text[829]

'yet another "apocolypse" show written by "big city dwellers" who have a total "city centered" view of everything.let me start by sayiing that the premise of the shows "disaster" is not too bad.  we are drawn in by being left guessing as to the nature of what has happened.this is where it wanders off the path.  the small town of jericho is nothing more than a minature version of new york or la from other movies.  with the exception of one or two "farms," everything ceases to exist at the town limits.  a character is either "of the town" or an "outsider" and that is it.it is plainly obvious that the writers have never spent any time in the real world, outside of a big city.  rural america does not work like this.  small towns blur.  the town limits rarely really limit the towns themselves and they often sort of blend into nearby communities and they really do encompass surrounding rural areas.  this is not europe and the local twons here are not, and never were, the local strongholds, c

In [130]:
## replace &#8217; back to Apostrophe Mark '
clean_text = [re.sub(r'&#39;|&#8216;|&#8217;', "'", sent) for sent in clean_text]

In [131]:
# text_ls[1176181]

In [132]:
# clean_text[1176181]

In [133]:
# text_ls[1175026]

In [134]:
# clean_text[1175026]

In [135]:
# replace &#8230; back to Horizontal Ellipsis …   .......remove
# clean_text = [re.sub(r'&#8230;', '…', sent) for sent in clean_text]

In [136]:
# text_ls[1341546]

In [137]:
# clean_text[1341546]

In [138]:
# text_ls[1627396]

In [139]:
# clean_text[1627396]

In [140]:
# replace &#8211; back to En Dash –    .......remove
# clean_text = [re.sub(r'&#8211;', "–", sent) for sent in clean_text]

In [141]:
# text_ls[1372525]

In [142]:
# clean_text[1372525]

In [143]:
# replace &#8212; back to Em Dash —    .......remove
# clean_text = [re.sub(r'&#8212;', "—", sent) for sent in clean_text]

In [144]:
# text_ls[1756309]

In [145]:
# clean_text[1756309]

In [146]:
# replace &#60; back to Less-Than Sign <    .......remove
# clean_text = [re.sub(r'&#60;', " < ", sent) for sent in clean_text]
# replace &#62; back to Greater-Than Sign >    .......remove
# clean_text = [re.sub(r'&#62;', " > ", sent) for sent in clean_text]

In [147]:
# text_ls[928474]

In [148]:
# clean_text[928474]

In [149]:
# text_ls[1949730]

In [150]:
# clean_text[1949730]

### replace star like symbol to <STAR>

In [151]:
## replace &#9733;  &#9734;  &#10032; back to <STAR> token, Sign ★
clean_text = [re.sub(r'&#9733;|&#9734;|&#10032;|&#10026;', " <STAR> ", sent) for sent in clean_text]

In [152]:
# text_ls[2050688]
# text_ls[2048172]
# text_ls[2024216]

In [153]:
# clean_text[2050688]
# clean_text[2048172]
# clean_text[2024216]

### replace :) as smiling face and :( as disappointed face

In [154]:
## replace :) as :smiling_face: ## 1350, 4246
clean_text = [re.sub(r'\:\)', " :smiling_face: ", sent) for sent in clean_text]

In [155]:
## replace :( as :disappointed_face:  ## 250, 712
clean_text = [re.sub(r'\:\(', " :disappointed_face: ", sent) for sent in clean_text]

### remove rest &#number

In [156]:
## remove all rest unicode decimal code &# 
clean_text = [re.sub(r'\&\#\d+\;{1}?', " ", sent) for sent in clean_text]

## 6.replace 1\*0%, 2\*0%, 10\*% to \<HUNDREDPERCENTAGE\>, rest remove

https://stackoverflow.com/questions/23202384/regular-expression-to-remove-all-numbers-and-all-dots

https://stackoverflow.com/questions/5110536/match-0-to-100-or-0-to-100-using-regex

In [157]:
sd, ds = find_noise(r'[1-2]{1}[0-9]{1}[0]{1}\s*%|[1]{1}[0]{1}[1-9]{1}\s*%', clean_text)  ## '  &lsquo;|&rsquo;|&apos;   "  &ldquo;|&rdquo;|&quot;     \w+\-\w+   \&\#\d+\;{1}?   \&\w+;
print(sd) ## 9373, 100%|100\s%|200%|200\s%  ## 100,200 #9373, [1-2]{1}[0]{2}\s*%    ## 1**, #9527, start with [1]{1}[0-9]{2}\s*%
    ## 1*0 2*0, # 9546  [1-2]{1}[0-9]{1}[0]{1}\s*%    ## 10* # [1]{1}[0]{1}[1-9]{1}\s*%
ds   ## Good one 9570, 1008， 3801 ## 1*0, 2*0, 10* # [1-2]{1}[0-9]{1}[0]{1}\s*%|[1]{1}[0]{1}[1-9]{1}\s*%

3801


[[940, ['100%']],
 [2827, ['100%', '100%']],
 [2904, ['100%', '100%']],
 [3197, ['100%']],
 [4192, ['100%']],
 [5003, ['100%']],
 [5749, ['100 %']],
 [5992, ['100%']],
 [6345, ['100%']],
 [6349, ['100%']],
 [6912, ['100%']],
 [7697, ['100%']],
 [7878, ['100%']],
 [7909, ['100%']],
 [8514, ['100%']],
 [9297, ['100%']],
 [10296, ['100%']],
 [12820, ['100%']],
 [14699, ['100%']],
 [15231, ['100%']],
 [16614, ['100%']],
 [17090, ['100%']],
 [17794, ['100%']],
 [18870, ['100%']],
 [19516, ['100%']],
 [20741, ['100%']],
 [20960, ['100%']],
 [21064, ['100%', '100%']],
 [21295, ['100%']],
 [21409, ['100%']],
 [21599, ['100%']],
 [22035, ['100%']],
 [23512, ['100%']],
 [23598, ['100%', '100%']],
 [24059, ['100%']],
 [24660, ['100%']],
 [25583, ['110%']],
 [27717, ['100%']],
 [28103, ['100%']],
 [28501, ['100%']],
 [29109, ['100%']],
 [31715, ['100%']],
 [31898, ['100%']],
 [31955, ['100%']],
 [32060, ['100%']],
 [32579, ['100%']],
 [32679, ['100%']],
 [32894, ['100%']],
 [33274, ['100%']],
 [33

In [158]:
# for i in ds:
#   print(clean_text[i[0]] + '\n')

In [159]:
# clean_text[2152277]

In [160]:
ss = []
for i in ds:
  for j in i[1]:
    ss.append(j)

In [161]:
from collections import Counter
print(len(Counter(ss).most_common()))
Counter(ss).most_common()

24


[('100%', 4182),
 ('120%', 52),
 ('100 %', 40),
 ('110%', 38),
 ('200%', 23),
 ('150%', 15),
 ('140%', 11),
 ('130%', 5),
 ('160%', 4),
 ('101%', 2),
 ('230%', 2),
 ('180 %', 2),
 ('260%', 2),
 ('103%', 1),
 ('200 %', 1),
 ('110 %', 1),
 ('180%', 1),
 ('270%', 1),
 ('240%', 1),
 ('250%', 1),
 ('105%', 1),
 ('170%', 1),
 ('108%', 1),
 ('280%', 1)]

### replace to \<HUNDREDPERCENTAGE\>

In [162]:
## replace to <HUNDRED_PERCENTAGE>
clean_text = [re.sub(r'[1-2]{1}[0-9]{1}[0]{1}\s*%|[1]{1}[0]{1}[1-9]{1}\s*%', " <HUNDREDPERCENTAGE> ", sent) for sent in clean_text]

In [163]:
sd, ds = find_noise(r'[1-2]{1}[0-9]{1}[0]{1}\s*%|[1]{1}[0]{1}[1-9]{1}\s*%', clean_text)  ## '  &lsquo;|&rsquo;|&apos;   "  &ldquo;|&rdquo;|&quot;     \w+\-\w+   \&\#\d+\;{1}?   \&\w+;
print(sd) ## 9373, 100%|100\s%|200%|200\s%  ## 100,200 #9373, [1-2]{1}[0]{2}\s*%    ## 1**, #9527, start with [1]{1}[0-9]{2}\s*%
    ## 1*0 2*0, # 9546  [1-2]{1}[0-9]{1}[0]{1}\s*%    ## 10* # [1]{1}[0]{1}[1-9]{1}\s*%
ds   ## Good one ## 1*0, 2*0, 10* # [1-2]{1}[0-9]{1}[0]{1}\s*%|[1]{1}[0]{1}[1-9]{1}\s*%

0


[]

### remove rest percentage

In [164]:
## good one
count_cent, found_cent = find_noise(r'\d+[\,\.\d+]*[\d+]*[\d+\s]*[\s]*%', clean_text) ## 29490, 21840  00,00 %
print(count_cent, len(found_cent))
## found 29490, 21840 percentage like 00% , 00 % , 00.00% , 00.00 % , 00,00% , 00,00 %
found_cent[:1000] ### 2035， 8640

8640 8640


[[305, ['90%']],
 [511, ['30%']],
 [702, ['95%', '5%']],
 [794, ['90%']],
 [821, ['99%']],
 [1075, ['5%']],
 [1167, ['10%']],
 [1272, ['70%']],
 [1509, ['50%']],
 [1975, ['90%']],
 [2077, ['95%']],
 [2315, ['99,5%']],
 [2560, ['1%']],
 [2710, ['95%']],
 [3644, ['10%']],
 [3660, ['10%']],
 [4035, ['99%']],
 [4372, ['99%']],
 [4375, ['99%']],
 [4406, ['95%']],
 [4445, ['90%']],
 [4499, ['95%']],
 [4669, ['80%']],
 [5276, ['75%', '1%', '1900%']],
 [5278, ['2. 75%', '1%', '4. 40%', '2%']],
 [5736, ['25%']],
 [5741, ['99%']],
 [5919, ['1%']],
 [6007, ['90%']],
 [6415, ['99%']],
 [6594, ['90%']],
 [6833, ['50%', '50%']],
 [6838, ['49%']],
 [6890, ['90%']],
 [6906, ['95%']],
 [7159, ['1%']],
 [7600, ['95%']],
 [7606, ['99%']],
 [7962, ['99.99%']],
 [7974, ['80 %']],
 [8166, ['10%', '85%']],
 [8532, ['68%']],
 [8711, ['75%']],
 [8948, ['80%', '20%']],
 [9066, ['50%']],
 [9149, ['1%']],
 [9151, ['1%']],
 [9262, ['300%']],
 [9388, ['30%']],
 [9412, ['1000%']],
 [9700, ['90%']],
 [9813, ['75%', '

In [ ]:
# clean_text[200735]

In [165]:
ss = []
for i in found_cent:
  for j in i[1]:
    ss.append(j)

In [166]:
from collections import Counter
print(len(Counter(ss).most_common()))
Counter(ss).most_common() ## 215, 492

492


[('50%', 1185),
 ('10%', 837),
 ('90%', 811),
 ('20%', 758),
 ('2%', 669),
 ('25%', 614),
 ('70%', 534),
 ('15%', 464),
 ('30%', 463),
 ('80%', 423),
 ('40%', 406),
 ('0%', 405),
 ('99%', 404),
 ('60%', 378),
 ('95%', 369),
 ('5%', 318),
 ('1%', 315),
 ('75%', 308),
 ('4%', 304),
 ('8%', 242),
 ('6%', 230),
 ('85%', 184),
 ('35%', 148),
 ('12%', 141),
 ('3%', 139),
 ('98%', 130),
 ('7%', 105),
 ('33%', 95),
 ('54%', 86),
 ('9%', 82),
 ('11%', 74),
 ('45%', 66),
 ('72%', 64),
 ('51%', 62),
 ('13%', 61),
 ('99.9%', 58),
 ('16%', 55),
 ('65%', 54),
 ('28%', 53),
 ('86%', 47),
 ('23%', 47),
 ('62%', 47),
 ('24%', 47),
 ('36%', 47),
 ('97%', 45),
 ('18%', 44),
 ('14%', 41),
 ('55%', 37),
 ('22%', 35),
 ('29%', 31),
 ('17%', 29),
 ('26%', 28),
 ('34%', 28),
 ('41%', 28),
 ('37%', 27),
 ('68%', 25),
 ('93%', 25),
 ('96%', 24),
 ('94%', 22),
 ('88%', 22),
 ('19%', 22),
 ('1.5%', 21),
 ('1000%', 20),
 ('66%', 20),
 ('78%', 19),
 ('47%', 19),
 ('400%', 18),
 ('0.5%', 18),
 ('32%', 17),
 ('71%', 

In [167]:
## replace rest to non <HUNDRED_PERCENTAGE>
clean_text = [re.sub(r'\d+[\,\.\d+]*[\d+]*[\d+\s]*[\s]*%', " ", sent) for sent in clean_text]

In [168]:
## good one
count_cent, found_cent = find_noise(r'\d+[\,\.\d+]*[\d+]*[\d+\s]*[\s]*%', clean_text) ## 29490  00,00 %
print(count_cent, len(found_cent))
## found 29490 percentage like 00% , 00 % , 00.00% , 00.00 % , 00,00% , 00,00 %
found_cent[:1000]

0 0


[]

## 7.remove numbers

In [169]:
## good one
count_cent, found_cent = find_noise(r'\d+', clean_text)
print(count_cent, len(found_cent))
found_cent[:1000] ## 91224, 199286

199286 199286


[[4, ['8']],
 [5, ['1']],
 [7, ['20', '30']],
 [23, ['30']],
 [27, ['2']],
 [28, ['5']],
 [31, ['2', '3', '4', '2', '4']],
 [34, ['2']],
 [36, ['4', '5', '7']],
 [37, ['4']],
 [38, ['4']],
 [41, ['3']],
 [42, ['2', '20']],
 [44, ['2005']],
 [49, ['1', '6']],
 [50, ['2001', '5', '3']],
 [51, ['24']],
 [53, ['5', '24', '24']],
 [57, ['24', '24', '24', '18', '24', '24', '5', '2006']],
 [58, ['24', '4', '2007']],
 [59, ['24', '6', '24']],
 [60, ['5', '1', '3']],
 [61, ['1', '4', '1', '2', '5']],
 [63, ['24']],
 [65, ['4', '5']],
 [66, ['24', '24', '24']],
 [67, ['24', '24']],
 [68, ['24', '1', '8']],
 [69, ['5', '5', '4', '5', '4', '4', '5', '24', '24', '24']],
 [70, ['24']],
 [73, ['24', '7', '24', '24', '6']],
 [74, ['24', '24', '9', '10']],
 [75, ['24', '4', '24']],
 [76, ['24', '45']],
 [77, ['1', '5', '5']],
 [78, ['24']],
 [79, ['24', '24', '24']],
 [80, ['24', '24', '2007', '24', '100']],
 [82, ['24', '2', '4', '24']],
 [90, ['5', '1991']],
 [93, ['4']],
 [96, ['3']],
 [98, ['3', '1

In [170]:
ss = []
for i in found_cent:
  for j in i[1]:
    ss.append(j)

In [171]:
from collections import Counter
print(len(Counter(ss).most_common()))
Counter(ss).most_common() # 4013, 7020

7020


[('2', 84452),
 ('1', 69075),
 ('3', 68489),
 ('5', 67774),
 ('4', 57614),
 ('10', 32965),
 ('8', 27435),
 ('6', 26610),
 ('12', 18660),
 ('7', 17198),
 ('20', 13591),
 ('9', 13295),
 ('50', 10718),
 ('30', 10133),
 ('15', 10121),
 ('11', 8244),
 ('16', 7891),
 ('100', 7827),
 ('0', 7097),
 ('80', 7055),
 ('18', 7039),
 ('40', 6907),
 ('60', 6087),
 ('13', 5795),
 ('90', 5781),
 ('25', 5591),
 ('24', 5534),
 ('70', 5308),
 ('14', 5237),
 ('17', 3104),
 ('45', 2818),
 ('200', 2764),
 ('19', 2587),
 ('21', 2494),
 ('22', 2428),
 ('000', 2341),
 ('300', 2294),
 ('2013', 2247),
 ('35', 2102),
 ('75', 2068),
 ('32', 2062),
 ('2012', 2037),
 ('500', 2029),
 ('2000', 1811),
 ('1970', 1806),
 ('120', 1798),
 ('1980', 1745),
 ('150', 1668),
 ('2011', 1542),
 ('36', 1518),
 ('23', 1456),
 ('1000', 1440),
 ('600', 1416),
 ('28', 1402),
 ('400', 1394),
 ('800', 1223),
 ('110', 1222),
 ('48', 1193),
 ('2010', 1177),
 ('2001', 1161),
 ('130', 1142),
 ('00', 1116),
 ('26', 1114),
 ('27', 1094),
 ('25

### replace numbers to empty

In [172]:
## replace numbers to 
clean_text = [re.sub(r'\d+', " ", sent) for sent in clean_text]

In [173]:
## good one
count_cent, found_cent = find_noise(r'\d+', clean_text)
print(count_cent, len(found_cent))
found_cent[:1000]

0 0


[]

In [174]:
## good one
count_cent, found_cent = find_noise(r'\d{4}', clean_text)
print(count_cent, len(found_cent))
found_cent[:1000]

0 0


[]

## 8.fix contraction

About [contractions](https://stackoverflow.com/questions/19790188/expanding-english-language-contractions-in-python)

https://medium.com/@lukei_3514/dealing-with-contractions-in-nlp-d6174300876b

In [175]:
!pip install contractions

     |████████████████████████████████| 245kB 5.8MB/s 
     |████████████████████████████████| 317kB 24.6MB/s 
  Created wheel for pyahocorasick: filename=pyahocorasick-1.4.0-cp36-cp36m-linux_x86_64.whl size=81695 sha256=f972a4b520ef4f4e8c99fe716dcf87c8227c42503ea39caf21f12161d68a97a1
  Stored in directory: /root/.cache/pip/wheels/0a/90/61/87a55f5b459792fbb2b7ba6b31721b06ff5cf6bde541b40994
Successfully built pyahocorasick


In [176]:
import contractions

## fix contractions for each sent
clean_text = [contractions.fix(sent) for sent in clean_text]

## release memory

In [177]:
%whos

Variable               Type             Data/Info
-------------------------------------------------
Counter                type             <class 'collections.Counter'>
UnknownEntities        type             <class 'namedentities.core.UnknownEntities'>
a                      list             n=2
cha                    list             n=0
clean_text             list             n=484826
contractions           module           <module 'contractions' fr<...>ontractions/__init__.py'>
converted              list             n=169
core                   module           <module 'namedentities.co<...>s/namedentities/core.py'>
count_cent             int              0
count_encode           int              4498
count_symbol           int              16408
count_url              int              542
decimal_entities       function         <function numeric_entities at 0x7f7dbf251f28>
df_review              DataFrame                                 <...>[484826 rows x 1 columns]
dict_unicod

In [178]:
## remove variables
del url_amazon, url_book, url_facebook, url_google, url_youtube, token_emoji, unicode_2_emoji, unicode_entities, found_url, found_encode

In [179]:
## remove variables 2
del found_key_url, converted, df_review, symbol, emoji_set, encode_el, html_entity_unique

## 9.clean symbol, punctuation (TBD)

- ('.', ' \<PERIOD\> ')
- (',', ' \<COMMA\> ')
- ('!', ' \<EXCLAMATION\> ')
- ('?', ' \<QUESTION\> ')
- ' \<APOSTROPHE\>
- <>


In [180]:
from string import punctuation

print('All punctuation: ', punctuation)
## keep ! ? , . ' < >     later deal with _
punctuation_2_remove = re.sub("[\!\?\,\.\'\<\>]", "", punctuation)
print('punctuation to remove : ', punctuation_2_remove)

All punctuation:  !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
punctuation to remove :  "#$%&()*+-/:;=@[\]^_`{|}~


### clean Non-need punctuation

- 511526 1 
- 1639671 2 
- 1821690 3 

In [181]:
# text_ls[511526], text_ls[1639671], text_ls[1821690]
# clean_text[511526], clean_text[1639671], clean_text[1821690]

In [182]:
### remove non-need punctuation, as ' '
for i, sent in enumerate(clean_text):
  clean_text[i] = ''.join([c if c not in punctuation_2_remove else ' ' for c in sent])

In [183]:
# clean_text[511526], clean_text[1639671], clean_text[1821690]

In [184]:
# clean_text[963498] ## lots of =====
# clean_text[152698] ## lots of $$$$
# clean_text[1437368] ## lots of ,,,,,,
# clean_text[2142476] ## =================
# clean_text[106] ## ---
# clean_text[708904] ## -----------
clean_text[57] ## --  -
# clean_text[2360363]

"for sheer intensity and plot topsy turvies, no other show comes close to  . you would think the premise  every hour of the   episode long season is played in real time  thus   encapsulates a full day  would get old by now, but, surprisingly, the producers and writers have managed to come up with another riveting, must see season. in fact, this might be the best season of them all. the thrills and nail biting suspense  as well as the spoilers alert  begin early in the season five opener, with the shocking slaying of president palmer and assassination attempts on familiar ctu agents  one of which is tragically successful . jack bauer  kiefer sutherland  is lured out of hiding after   months and becomes rapidly embroiled in a devious conspiracy that goes all the way up to the white house. as he becomes the most wanted man in america  he is framed for palmer's assassination , he must try to ferret out the true culprit s  and again save the country from deadly russian extremists and variou

In [185]:
clean_text[63676]

' pac i mean makaveli really took his time with these songs on the album. hail mary had one of the coldest beats known to man. none of the songs were just ok or good on this album. all of them were perfect. r i p  pac aka makaveli'

In [186]:
text_ls[63676]

'2pac I mean Makaveli really took his time with these songs on the album. Hail Mary had one of the coldest beats known to man. None of the songs were just ok or good on this album. all of them were perfect. R I P 2pac aka Makaveli'

In [187]:
clean_text[9568]

"a man phones a lawyer and asks,  how much would you charge for just answering three simple questions? the lawyer replies,  a thousand dollars.  a thousand dollars!  exclaims the man.  that is very expensive is not it?  it certainly is,  says the lawyer.  now, what is your third question? or this one...q  what do you have if three lawyers are buried up to their necks in cement?a  not enough cement.the usa network keeps on not effing up. far as I am concerned, suits is the newest addictive show on television, and that is me not normally drawn to legal dramas. it is witty and funny and smart and loaded with that patented easy breezy sense of fun that the usa network nowadays specializes in. the premise is irresistible, the cast is irrepressible. the two leads, gabriel macht and patrick j. adams, are so very good. i can not not rave.suits chronicles the exploits of the prestigious new york corporate law firm pearson hardman. more pointedly, the catchy premise revolves around sleek newly m

In [188]:
text_ls[9568]

'A man phones a lawyer and asks, "How much would you charge for just answering three simple questions?"The lawyer replies, "A thousand dollars.""A thousand dollars!" exclaims the man. "That\'s very expensive isn\'t it?""It certainly is," says the lawyer. "Now, what\'s your third question?"Or this one...Q: What do you have if three lawyers are buried up to their necks in cement?A: Not enough cement.The USA network keeps on not effing up. Far as I\'m concerned, SUITS is the newest addictive show on television, and that\'s me not normally drawn to legal dramas. It\'s witty and funny and smart and loaded with that patented easy-breezy sense of fun that the USA network nowadays specializes in. The premise is irresistible, the cast is irrepressible. The two leads, Gabriel Macht and Patrick J. Adams, are so very good. I can\'t not rave.SUITS chronicles the exploits of the prestigious New York corporate law firm Pearson Hardman. More pointedly, the catchy premise revolves around sleek newly-ma

### keep for each \!\?\,\.\'\<\> only one

In [189]:
# \.{2,}  ('.', ' \<PERIOD\> ')
# \,{2,}  (',', ' \<COMMA\> ')
# \!{2,}  ('!', ' \<EXCLAMATION\> ')
# \?{2,}  ('?', ' \<QUESTION\> ')
# \'{1,}  ("\'", ' \<APOSTROPHE\> ')
# keep <>
clean_text = [re.sub(r'\.{1,}', " <PERIOD> ", sent) for sent in clean_text]
clean_text = [re.sub(r'\,{1,}', " <COMMA> ", sent) for sent in clean_text] 
clean_text = [re.sub(r'\!{1,}', " <EXCLAMATION> ", sent) for sent in clean_text]
clean_text = [re.sub(r'\?{1,}', " <QUESTION> ", sent) for sent in clean_text]

In [190]:
clean_text = [re.sub(r'\'{1,}', "\'", sent) for sent in clean_text]   # <APOSTROPHE> 

In [191]:
clean_text = [re.sub(r"\'s", " \'s", sent) for sent in clean_text]
clean_text = [re.sub(r"\'ve", " \'ve", sent) for sent in clean_text]
clean_text = [re.sub(r"n\'t", " n\'t", sent) for sent in clean_text] # NOTCONTRACT
clean_text = [re.sub(r"\'re", " \'re", sent) for sent in clean_text]
clean_text = [re.sub(r"\'d", " \'d", sent) for sent in clean_text]
clean_text = [re.sub(r"\'ll", " \'ll", sent) for sent in clean_text]

In [192]:
clean_text = [re.sub(r'n\'t', ' <NOTCONTRACT> ', sent) for sent in clean_text]

In [193]:
## remove ' 
clean_text = [re.sub(r"\'", ' ', sent) for sent in clean_text]

### check pattern before remove stopword

In [198]:
## good one  keep ! ? need to remove : + ;
count_cent, found_cent = find_noise(r"<NOTCONTRACT>", clean_text) ## 294405 \s\w+\-\w+\s    23314  \s\w+\--\w+\s
print(count_cent, len(found_cent))  ## 184152  \s\w+\-{1,2}\w+\-{1,2}\w+\s|\s\w+\-{1,2}\w+\s   ## 1  152423   ## 2  31692 don't care
found_cent[:1000]       ## 3869, 579, 921 \s<NOTCONTRACT>\s     ## 409210  \w+\'\w+   1287756            ## \s\w+\-{1}\w+\-{1}\w+\s|\s\w+\-{1}\w+\s

921 921


[[1780, ['<NOTCONTRACT>']],
 [1934, ['<NOTCONTRACT>']],
 [1962, ['<NOTCONTRACT>']],
 [2110, ['<NOTCONTRACT>']],
 [3624, ['<NOTCONTRACT>']],
 [6032, ['<NOTCONTRACT>']],
 [6739, ['<NOTCONTRACT>']],
 [7515, ['<NOTCONTRACT>']],
 [8091, ['<NOTCONTRACT>']],
 [9780, ['<NOTCONTRACT>']],
 [12906, ['<NOTCONTRACT>']],
 [15675, ['<NOTCONTRACT>']],
 [16486, ['<NOTCONTRACT>']],
 [16847, ['<NOTCONTRACT>']],
 [17717, ['<NOTCONTRACT>']],
 [19477, ['<NOTCONTRACT>']],
 [20501, ['<NOTCONTRACT>']],
 [21333, ['<NOTCONTRACT>']],
 [21941, ['<NOTCONTRACT>']],
 [23604, ['<NOTCONTRACT>']],
 [23967, ['<NOTCONTRACT>']],
 [23975, ['<NOTCONTRACT>']],
 [24757, ['<NOTCONTRACT>']],
 [27145, ['<NOTCONTRACT>']],
 [30600, ['<NOTCONTRACT>']],
 [30687, ['<NOTCONTRACT>']],
 [32272, ['<NOTCONTRACT>']],
 [32298, ['<NOTCONTRACT>']],
 [32404, ['<NOTCONTRACT>']],
 [33194, ['<NOTCONTRACT>']],
 [35010, ['<NOTCONTRACT>']],
 [35364, ['<NOTCONTRACT>']],
 [35541, ['<NOTCONTRACT>']],
 [35590, ['<NOTCONTRACT>']],
 [36007, ['<NOTCONTRACT>

In [200]:
ss = []
for i in found_cent:
  for j in i[1]:
    ss.append(j)

from collections import Counter 
print(len(Counter(ss).most_common()))
Counter(ss).most_common() # clean 48079, 638, 986 # text_ls  58203

1


[('<NOTCONTRACT>', 986)]

In [201]:
%whos

Variable               Type             Data/Info
-------------------------------------------------
Counter                type             <class 'collections.Counter'>
UnknownEntities        type             <class 'namedentities.core.UnknownEntities'>
a                      list             n=2
cha                    list             n=0
clean_text             list             n=484826
contractions           module           <module 'contractions' fr<...>ontractions/__init__.py'>
core                   module           <module 'namedentities.co<...>s/namedentities/core.py'>
count_cent             int              921
count_encode           int              4498
count_symbol           int              16408
count_url              int              542
decimal_entities       function         <function numeric_entities at 0x7f7dbf251f28>
dict_unicode_2_emoji   dict             n=157
drive                  module           <module 'google.colab.dri<...>s/google/colab/drive.py'>
ds       

In [202]:
del count_cent, count_encode, count_symbol, count_url, a, cha, dict_unicode_2_emoji, found_cent, ss, find_noise_url, contractions, hex_entities, html, named_entities, numeric_entities, punctuation, punctuation_2_remove, unescape

In [203]:
del ds, el, emoji, encode_ampersands, entities, i, j, key, sd, UnknownEntities

In [204]:
!ps -aux|grep python

root          49  0.2  0.4 196204 62636 ?        Sl   12:40   0:05 /usr/bin/python2 /usr/local/bin/jupyter-notebook --ip="172.28.0.2" --port=9000 --FileContentsManager.root_dir="/" --LargeFileManager.delete_to_trash=False --MappingKernelManager.root_dir="/content"
root          56 15.7 12.8 2120760 1709804 ?     Ssl  12:40   5:15 /usr/bin/python3 -m ipykernel_launcher -f /root/.local/share/jupyter/runtime/kernel-b396b8a6-ed73-4e1f-a8dd-cb316ab1875a.json
root         244  0.0  0.0  18384  3020 ?        S    12:41   0:00 bash -c tail -n +0 -F "/root/.config/Google/DriveFS/Logs/drive_fs.txt" | python3 /opt/google/drive/drive-filter.py > "/root/.config/Google/DriveFS/Logs/timeouts.txt" 
root         246  0.0  0.0  29216  9376 ?        S    12:41   0:00 python3 /opt/google/drive/drive-filter.py
root         659  0.0  0.0  39200  6568 ?        S    13:13   0:00 /bin/bash -c ps -aux|grep python
root         661  0.0  0.0  38580  5028 ?        S    13:13   0:00 grep python


In [205]:
# !kill -9 

### Stopwords removal NLTK

https://www.geeksforgeeks.org/removing-stop-words-nltk-python/?ref=lbp

https://www.kite.com/python/docs/nltk.word_tokenize

https://stackabuse.com/removing-stop-words-from-strings-in-python/#usingthespacylibrary


https://medium.com/towards-artificial-intelligence/stop-the-stopwords-using-different-python-libraries-ffa6df941653

In [206]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [207]:
nltk.download('punkt')
from nltk.tokenize import word_tokenize
print(word_tokenize("dosn't does't arn't dosen't d!a 'harry potter' g'aould o'brian havn't nick's ad ? sad. sada,Hedon't"))

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
['dos', "n't", "does't", 'ar', "n't", 'dose', "n't", 'd', '!', 'a', "'harry", 'potter', "'", "g'aould", "o'brian", 'hav', "n't", 'nick', "'s", 'ad', '?', 'sad', '.', 'sada', ',', 'Hedo', "n't"]


In [208]:
# [i for i in tokenizer.tokenize(txt) if (i not in stopwords_list or i in whitelist)]

In [209]:
from nltk.corpus import stopwords
print(len(stopwords.words('english')))
print(stopwords.words('english'))
# stopwords.words('english')

179
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than

In [210]:
from nltk.corpus import stopwords

stopwords_list = stopwords.words('english')
whitelist = ["not", "no", "just", "so"]

### create word_counts dictionary for all words

In [211]:
word_counts = []
for row in clean_text:
  for i in row.split():
    # print(i)
    word_counts.append(i)

In [212]:
from collections import Counter 

print('Total number of words in cleaned corpus:', len(word_counts))
print('Number of unique words in cleaned corpus:', len(Counter(word_counts)), len(Counter(word_counts).most_common()))

Total number of words in cleaned corpus: 65163194
Number of unique words in cleaned corpus: 222990 222990


6 classes

Total number of words in cleaned corpus: 32574069

Number of unique words in cleaned corpus: 159129 159129


8 classes

Total number of words in cleaned corpus: 65163194

Number of unique words in cleaned corpus: 222990 222990

In [213]:
# get count of each word
from collections import Counter 
word_counts = Counter(word_counts)
# Counter(word_counts).most_common()

In [214]:
word_counts.most_common(5)

[('<PERIOD>', 3476956),
 ('the', 2940758),
 ('<COMMA>', 2399259),
 ('and', 1621711),
 ('a', 1607055)]

In [215]:
import json

with open(root_path + '/word_counts.json', 'w') as fp:
    json.dump(word_counts, fp)

## load word_counts dictionary

In [216]:
import json

with open(root_path + '/word_counts.json', 'r') as fp:
    word_counts = json.load(fp)

print(len(word_counts.keys()))
# word_counts

222990


In [217]:
# "what this is just not".split()
[word for word in "what this is just not".split() if (word not in stopwords_list or word in whitelist) and len(word) > 1 and word_counts[word] > 5]

['just', 'not']

### for each sent, remove stopwords & word len <= 1 & word_counts <=5

In [218]:
# Remove all words with  5 or fewer occurences
for i, sent in enumerate(clean_text):
  clean_text[i] = [word for word in sent.split() if (word not in stopwords_list or word in whitelist) and len(word) > 1 and word_counts[word] > 5]

In [219]:
print(text_ls[0])
print(clean_text[0])
print(" ".join(clean_text[0]))

I had big expectations because I love English TV, in particular Investigative and detective stuff but this guy is really boring. It didn't appeal to me at all.
['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>']
big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>


In [220]:
%whos

Variable           Type                    Data/Info
----------------------------------------------------
Counter            type                    <class 'collections.Counter'>
clean_text         list                    n=484826
core               module                  <module 'namedentities.co<...>s/namedentities/core.py'>
decimal_entities   function                <function numeric_entities at 0x7f7dbf251f28>
drive              module                  <module 'google.colab.dri<...>s/google/colab/drive.py'>
file               TextIOWrapper           <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
find_noise         function                <function find_noise at 0x7f7dc4500620>
fp                 TextIOWrapper           <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
i                  int                     484825
json               module                  <module 'json' from '/usr<...>hon3.6/json/__init__.py'>
line               str                     Very nice

In [221]:
del stopwords, nltk, word_tokenize

### For empty review or unknown word in corpus replace as \<UNK\>

In [222]:
empty_ls = [] ## 10 classes dataset 224 empty row before remove stopword ## after remove stopword 279 
count = 0 ## finally 54, 138 empty rows after remove word_counts <= 5
for i, sent in enumerate(clean_text):
  if not len(sent):
    count += 1
    empty_ls.append(i)
    print(i, count, len(sent), sent, text_ls[i])

37144 1 0 [] 
39355 2 0 [] 
43811 3 0 [] 
44732 4 0 [] 
47060 5 0 [] 
48565 6 0 [] 
62118 7 0 [] 
86150 8 0 [] OMPHHHHHHH
106932 9 0 [] 7 OF 10
112128 10 0 [] 10 OF 10
125232 11 0 [] 
126706 12 0 [] 
131044 13 0 [] 
131480 14 0 [] 
131611 15 0 [] 
131618 16 0 [] 
131647 17 0 [] 
132735 18 0 [] 
138989 19 0 [] 
139150 20 0 [] 
140092 21 0 [] 
144603 22 0 [] 
145472 23 0 [] 
148315 24 0 [] 
148866 25 0 [] 
151052 26 0 [] 
151309 27 0 [] 
157893 28 0 [] 
171188 29 0 [] 
171728 30 0 [] 
172095 31 0 [] 
172913 32 0 [] 
175690 33 0 [] 
176030 34 0 [] 
176372 35 0 [] 
179210 36 0 [] 
179887 37 0 [] 
181931 38 0 [] 
185952 39 0 [] 
185999 40 0 [] 
186038 41 0 [] 
186472 42 0 [] 
186616 43 0 [] 
188283 44 0 [] 
188783 45 0 [] 
191018 46 0 [] 
194903 47 0 [] 
194952 48 0 [] 
195138 49 0 [] 
195933 50 0 [] 
196012 51 0 [] 
198906 52 0 [] 
202626 53 0 [] 
204499 54 0 [] 
206474 55 0 [] 
210796 56 0 [] 
217722 57 0 [] 
217950 58 0 [] 
223092 59 0 [] 
225628 60 0 [] 
234144 61 0 [] 
246050 62 0 [] 


In [223]:
len(empty_ls) ## 54

138

In [224]:
clean_text = [sent if sent else ' <UNK> ' for sent in clean_text]

In [225]:
for i in empty_ls:
  print(i, clean_text[i], text_ls[i])

37144  <UNK>  
39355  <UNK>  
43811  <UNK>  
44732  <UNK>  
47060  <UNK>  
48565  <UNK>  
62118  <UNK>  
86150  <UNK>  OMPHHHHHHH
106932  <UNK>  7 OF 10
112128  <UNK>  10 OF 10
125232  <UNK>  
126706  <UNK>  
131044  <UNK>  
131480  <UNK>  
131611  <UNK>  
131618  <UNK>  
131647  <UNK>  
132735  <UNK>  
138989  <UNK>  
139150  <UNK>  
140092  <UNK>  
144603  <UNK>  
145472  <UNK>  
148315  <UNK>  
148866  <UNK>  
151052  <UNK>  
151309  <UNK>  
157893  <UNK>  
171188  <UNK>  
171728  <UNK>  
172095  <UNK>  
172913  <UNK>  
175690  <UNK>  
176030  <UNK>  
176372  <UNK>  
179210  <UNK>  
179887  <UNK>  
181931  <UNK>  
185952  <UNK>  
185999  <UNK>  
186038  <UNK>  
186472  <UNK>  
186616  <UNK>  
188283  <UNK>  
188783  <UNK>  
191018  <UNK>  
194903  <UNK>  
194952  <UNK>  
195138  <UNK>  
195933  <UNK>  
196012  <UNK>  
198906  <UNK>  
202626  <UNK>  
204499  <UNK>  
206474  <UNK>  
210796  <UNK>  
217722  <UNK>  
217950  <UNK>  
223092  <UNK>  
225628  <UNK>  
234144  <UNK>  
246050 

In [226]:
del empty_ls, count, stopwords_list

In [227]:
empty_ls = [] ## 224 empty row before remove stopword ## after remove stopword 279 
count = 0 ## finally 54 empty rows after remove word_counts <= 5
for i, sent in enumerate(clean_text):
  if sent == " <UNK> ":
    count += 1
    empty_ls.append(i)
    print(i, count, len(sent), sent, text_ls[i])

37144 1 7  <UNK>  
39355 2 7  <UNK>  
43811 3 7  <UNK>  
44732 4 7  <UNK>  
47060 5 7  <UNK>  
48565 6 7  <UNK>  
62118 7 7  <UNK>  
86150 8 7  <UNK>  OMPHHHHHHH
106932 9 7  <UNK>  7 OF 10
112128 10 7  <UNK>  10 OF 10
125232 11 7  <UNK>  
126706 12 7  <UNK>  
131044 13 7  <UNK>  
131480 14 7  <UNK>  
131611 15 7  <UNK>  
131618 16 7  <UNK>  
131647 17 7  <UNK>  
132735 18 7  <UNK>  
138989 19 7  <UNK>  
139150 20 7  <UNK>  
140092 21 7  <UNK>  
144603 22 7  <UNK>  
145472 23 7  <UNK>  
148315 24 7  <UNK>  
148866 25 7  <UNK>  
151052 26 7  <UNK>  
151309 27 7  <UNK>  
157893 28 7  <UNK>  
171188 29 7  <UNK>  
171728 30 7  <UNK>  
172095 31 7  <UNK>  
172913 32 7  <UNK>  
175690 33 7  <UNK>  
176030 34 7  <UNK>  
176372 35 7  <UNK>  
179210 36 7  <UNK>  
179887 37 7  <UNK>  
181931 38 7  <UNK>  
185952 39 7  <UNK>  
185999 40 7  <UNK>  
186038 41 7  <UNK>  
186472 42 7  <UNK>  
186616 43 7  <UNK>  
188283 44 7  <UNK>  
188783 45 7  <UNK>  
191018 46 7  <UNK>  
194903 47 7  <UNK>  
19495

In [228]:
len(empty_ls)

138

In [229]:
len(clean_text)

484826

## 10.Save cleaned review text

In [230]:
# write into text file
with open('cleaned_reviews_text.txt', 'w') as f:
    for sent in clean_text:
      if sent==" <UNK> ":
        f.write("%s\n" % sent)
      else:
        f.write("%s\n" % " ".join(sent))

In [231]:
## Compress review text into zip file
!zip "{root_path}"/cleaned_reviews_text.zip cleaned_reviews_text.txt

updating: cleaned_reviews_text.txt (deflated 70%)


In [232]:
# clean_text = [sent if sent=="<UNK>" else " ".join(sent) for sent in clean_text]

### Read file and check data

In [234]:
!unzip "{root_path}"/cleaned_reviews_text.zip

Archive:  /content/drive/My Drive/Colab/MSC_RD/cleaned_reviews_text.zip
replace cleaned_reviews_text.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: A
  inflating: cleaned_reviews_text.txt  


In [235]:
handle_list = []
with open('/content/cleaned_reviews_text.txt', 'r') as file:
  for line in file:
    handle_list.append(line)

In [236]:
print(len(handle_list))
handle_list[1]

484826


'highly recommend series <PERIOD> must anyone yearning watch grown television <PERIOD> complex characters plots keep one totally involved <PERIOD> thank amazin prime <PERIOD>\n'

In [237]:
cleaned_text_ls = map(lambda s: s.strip('\n'), handle_list)
cleaned_text_ls = list(cleaned_text_ls)
cleaned_text_ls[:5]

['big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>',
 'highly recommend series <PERIOD> must anyone yearning watch grown television <PERIOD> complex characters plots keep one totally involved <PERIOD> thank amazin prime <PERIOD>',
 'one real snoozer <PERIOD> not believe anything read hear <COMMA> awful <PERIOD> no idea title means <PERIOD> neither <PERIOD>',
 'mysteries interesting <PERIOD> tension robson tall blond good not always believable <PERIOD> often seemed uncomfortable <PERIOD>',
 'show always excellent <COMMA> far british crime mystery one best ever made <PERIOD> stories well done acting top notch interesting twists realistic brutal storylines <PERIOD> show pulls no punches enters twisted minds criminals profiler psychiatrist helps northern english city police force <PERIOD> show looks like shot manchester called another name show <PERIOD> one episode not disc excellent prayer bone seperate disc <

In [238]:
# cleaned_text_ls[2360363]

In [239]:
cleaned_text_ls[9568]

'man phones lawyer asks <COMMA> much would charge just answering three simple questions <QUESTION> lawyer replies <COMMA> thousand dollars <PERIOD> thousand dollars <EXCLAMATION> exclaims man <PERIOD> expensive not <QUESTION> certainly <COMMA> says lawyer <PERIOD> <COMMA> third question <QUESTION> one <PERIOD> three lawyers buried necks cement <QUESTION> not enough cement <PERIOD> usa network keeps not effing <PERIOD> far concerned <COMMA> suits newest addictive show television <COMMA> not normally drawn legal dramas <PERIOD> witty funny smart loaded patented easy breezy sense fun usa network nowadays specializes <PERIOD> premise irresistible <COMMA> cast irrepressible <PERIOD> two leads <COMMA> gabriel macht patrick <PERIOD> adams <COMMA> so good <PERIOD> not not rave <PERIOD> suits chronicles exploits prestigious new york corporate law firm pearson hardman <PERIOD> pointedly <COMMA> catchy premise revolves around sleek newly made senior partner harvey specter macht dips shady shenani

In [240]:
# cleaned_text_ls[511526], cleaned_text_ls[1639671], cleaned_text_ls[1821690]

In [241]:
def find_noise(reg, datalist):
  count = 0
  found_list = []

  for i, row in enumerate(datalist):
    found = re.findall(reg, row)

    if found != []:
      count += 1
      # print(found[0])
      # found = list(filter(None, found[0]))
      # print(found)
      found_list.append([i ,found])
  
  return count, found_list
  # return found_list

In [242]:
import re
## good one  keep ! ? need to remove : + ;
count_key, found_key = find_noise(r" <UNK> ", cleaned_text_ls) ## 294405 \s\w+\-\w+\s    23314  \s\w+\--\w+\s
print(count_key, len(found_key))  ## 184152  \s\w+\-{1,2}\w+\-{1,2}\w+\s|\s\w+\-{1,2}\w+\s   ## 1  152423   ## 2  31692 don't care
found_key[:1000]       ## 3869 \s<NOTCONTRACT>\s     ## 409210  \w+\'\w+   1287756            ## \s\w+\-{1}\w+\-{1}\w+\s|\s\w+\-{1}\w+\s

138 138


[[37144, [' <UNK> ']],
 [39355, [' <UNK> ']],
 [43811, [' <UNK> ']],
 [44732, [' <UNK> ']],
 [47060, [' <UNK> ']],
 [48565, [' <UNK> ']],
 [62118, [' <UNK> ']],
 [86150, [' <UNK> ']],
 [106932, [' <UNK> ']],
 [112128, [' <UNK> ']],
 [125232, [' <UNK> ']],
 [126706, [' <UNK> ']],
 [131044, [' <UNK> ']],
 [131480, [' <UNK> ']],
 [131611, [' <UNK> ']],
 [131618, [' <UNK> ']],
 [131647, [' <UNK> ']],
 [132735, [' <UNK> ']],
 [138989, [' <UNK> ']],
 [139150, [' <UNK> ']],
 [140092, [' <UNK> ']],
 [144603, [' <UNK> ']],
 [145472, [' <UNK> ']],
 [148315, [' <UNK> ']],
 [148866, [' <UNK> ']],
 [151052, [' <UNK> ']],
 [151309, [' <UNK> ']],
 [157893, [' <UNK> ']],
 [171188, [' <UNK> ']],
 [171728, [' <UNK> ']],
 [172095, [' <UNK> ']],
 [172913, [' <UNK> ']],
 [175690, [' <UNK> ']],
 [176030, [' <UNK> ']],
 [176372, [' <UNK> ']],
 [179210, [' <UNK> ']],
 [179887, [' <UNK> ']],
 [181931, [' <UNK> ']],
 [185952, [' <UNK> ']],
 [185999, [' <UNK> ']],
 [186038, [' <UNK> ']],
 [186472, [' <UNK> ']],


In [243]:
ss = []
for i in found_key:
  for j in i[1]:
    ss.append(j)

from collections import Counter 
print(len(Counter(ss).most_common()))
Counter(ss).most_common() # clean 48079 # text_ls  58203

1


[(' <UNK> ', 138)]

In [244]:
for i in found_key:
  print(cleaned_text_ls[i[0]], text_ls[i[0]])

 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  OMPHHHHHHH
 <UNK>  7 OF 10
 <UNK>  10 OF 10
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  It worke
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UNK>  
 <UN

## replace

In [245]:
## pass url 2nd
# count_url, found_url = find_noise(r"http.?://[^\s]+[\s]?|www\.[^\s|\.|\"].?[^\s\"\>\)]+|youtube\.com[^\s\)]+", clean_text)

## pass url
# count_url, found_url = find_noise(r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))", clean_text)

## pass email
# count_url, found_url = find_noise(r'[\w\.-]+@[\w\.-]+\.\w+|[\w\.-]+@[\w\.-]+', clean_text)
## pass @ mention
# count_url, found_url = find_noise(r'@\w+', clean_text)
## pass dollor $ price
# count_url, found_url = find_noise(r'\$\d+(?:\,\d+|\.\d+)?', clean_text)
## lots of =====
# count_url, found_url = find_noise(r'(\==+)+?', clean_text)
# count_url, found_url = find_noise(r'(\$\$+)+?', clean_text) 
# count_url, found_url = find_noise(r'(\,,+)+?', clean_text) 
# count_url, found_url = find_noise(r'(\---+)+?', clean_text)

## match a single punctuation ## https://stackoverflow.com/questions/44524165/remove-punctuation-in-python-but-keep-emoticons
# count_url, found_url = find_noise(r'(?<=\w)[^\s\w](?![^\s\w])', clean_text) 

# <QUOTATION_MARK> "
# count_url, found_url = find_noise(r'&#34;', clean_text) 
# Right Single Quotation Mark '
# count_url, found_url = find_noise(r'&#8217;', clean_text) 
# Right Double Quotation Mark ”
# count_url, found_url = find_noise(r'&#8221;', clean_text) 
# Left Double Quotation Mark “
# count_url, found_url = find_noise(r'&#8220;', clean_text) 
# Horizontal Ellipsis …
# count_url, found_url = find_noise(r'&#8230;', clean_text) 

# Left Single Quotation Mark ‘
# count_url, found_url = find_noise(r'&#8216;', clean_text)

# En Dash – 7009
# count_url, found_url = find_noise(r'&#8211;', clean_text)  
# Em Dash — 2499
# count_url, found_url = find_noise(r'&#8212;', clean_text) 
# Less-Than Sign <
# count_url, found_url = find_noise(r'&#60;', clean_text) 
# Greater-Than Sign >
# count_url, found_url = find_noise(r'&#62;', clean_text) 
# HYPHEN -    .......remove
# count_url, found_url = find_noise(r'\-{1,2}', clean_text) 

# count_url, found_url = find_noise(r'=', clean_text) 

## find encode
count_url, found_url = find_noise(r'\&\#\d+\;{1}?', cleaned_text_ls) 

In [246]:
# clean_text[963498] ## lots of =====
# clean_text[152698] ## lots of $$$$
# clean_text[1437368] ## lots of ,,,,,,
# clean_text[2142476] ## =================
# clean_text[106] ## ---
# clean_text[708904] ## -----------
# clean_text[57] ## --  -
clean_text[41931]

['thing',
 'works',
 'great',
 'easy',
 'install',
 '<PERIOD>',
 'rv',
 'black',
 'grey',
 'tanks',
 '<PERIOD>']

In [247]:
print(count_url, len(found_url))
found_url

0 0


[]

## check tokens

In [248]:
count_url, found_url = find_noise(r'(\<\w+\>)+?', cleaned_text_ls) ##50499

In [249]:
print(count_url, len(found_url))
found_url[40000:] ## tokens in total 197723

481378 481378


[[40465,
  ['<COMMA>',
   '<PERIOD>',
   '<COMMA>',
   '<HUNDREDPERCENTAGE>',
   '<PERIOD>',
   '<COMMA>',
   '<PERIOD>',
   '<PERIOD>',
   '<PERIOD>',
   '<COMMA>',
   '<PERIOD>',
   '<PERIOD>',
   '<PERIOD>']],
 [40466,
  ['<COMMA>',
   '<COMMA>',
   '<PERIOD>',
   '<PERIOD>',
   '<PERIOD>',
   '<COMMA>',
   '<PERIOD>']],
 [40467,
  ['<PERIOD>',
   '<COMMA>',
   '<COMMA>',
   '<PERIOD>',
   '<COMMA>',
   '<COMMA>',
   '<PERIOD>',
   '<PERIOD>']],
 [40468,
  ['<PERIOD>',
   '<PERIOD>',
   '<COMMA>',
   '<PERIOD>',
   '<PERIOD>',
   '<PERIOD>',
   '<EXCLAMATION>',
   '<PERIOD>']],
 [40469, ['<COMMA>', '<PERIOD>', '<PERIOD>']],
 [40470,
  ['<PERIOD>',
   '<PERIOD>',
   '<PERIOD>',
   '<PERIOD>',
   '<PERIOD>',
   '<EXCLAMATION>',
   '<PERIOD>']],
 [40471, ['<PERIOD>', '<PERIOD>']],
 [40472, ['<COMMA>', '<PERIOD>', '<PERIOD>', '<PERIOD>']],
 [40473, ['<COMMA>', '<PERIOD>', '<EXCLAMATION>', '<COMMA>', '<PERIOD>']],
 [40474, ['<EXCLAMATION>', '<EXCLAMATION>', '<COMMA>', '<EXCLAMATION>']],


In [250]:
ss = []
for i in found_url:
  for j in i[1]:
    ss.append(j)

In [251]:
from collections import Counter
print(len(Counter(ss).most_common()))
Counter(ss).most_common() # 12， 14 kinds of token

14


[('<PERIOD>', 3476956),
 ('<COMMA>', 2399259),
 ('<EXCLAMATION>', 211162),
 ('<QUESTION>', 64627),
 ('<PRICE>', 29257),
 ('<HUNDREDPERCENTAGE>', 4389),
 ('<NOTCONTRACT>', 986),
 ('<AMAZONURL>', 506),
 ('<MENTION>', 288),
 ('<URL>', 246),
 ('<UNK>', 138),
 ('<YOUTUBEURL>', 25),
 ('<GOOGLEURL>', 13),
 ('<STAR>', 9)]

In [252]:
print(text_ls[14441])
print(" ".join(clean_text[14441]))
print(cleaned_text_ls[14441])

My husband and I enjoy the &#34;reality documentary&#34; programs, and this is a good one.  The officers are dedicated, polite, and family-oriented.  But they are also tough.  LA is an unusual state, and this program offers a different perspective on law enforcement on the bayou.
husband enjoy reality documentary programs <COMMA> good one <PERIOD> officers dedicated <COMMA> polite <COMMA> family oriented <PERIOD> also tough <PERIOD> la unusual state <COMMA> program offers different perspective law enforcement bayou <PERIOD>
husband enjoy reality documentary programs <COMMA> good one <PERIOD> officers dedicated <COMMA> polite <COMMA> family oriented <PERIOD> also tough <PERIOD> la unusual state <COMMA> program offers different perspective law enforcement bayou <PERIOD>


In [253]:
print(len(text_ls), len(clean_text), len(cleaned_text_ls))

484826 484826 484826


In [254]:
for i in found_key:
  print("".join(clean_text[i[0]]), text_ls[i[0]])
  print(cleaned_text_ls[i[0]])
  print()

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  OMPHHHHHHH
 <UNK> 

 <UNK>  7 OF 10
 <UNK> 

 <UNK>  10 OF 10
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <UNK>  
 <UNK> 

 <U

In [265]:
print(text_ls[0])
print(clean_text[0])
print(cleaned_text_ls[0])

I had big expectations because I love English TV, in particular Investigative and detective stuff but this guy is really boring. It didn't appeal to me at all.
['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>']
big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>


In [266]:
reviews_text_merge = " ".join(cleaned_text_ls)

In [267]:
words = reviews_text_merge.split()

In [268]:
word_counts = Counter(words)
print('Total number of words in cleaned corpus:', len(words))
print('Number of unique words in cleaned corpus:', len(Counter(words)), len(Counter(words).most_common()))

Total number of words in cleaned corpus: 36436399
Number of unique words in cleaned corpus: 55984 55984




https://stackoverflow.com/questions/44865840/how-to-treat-numbers-inside-text-strings-when-vectorizing-words

https://www.w3schools.com/python/python_regex.asp#findall

https://www.rexegg.com/regex-quickstart.html

https://stackoverflow.com/questions/46931715/missing-word-in-word-embedding

https://stackoverflow.com/questions/53837088/embedding-vs-inserting-word-vectors-directly-to-input-layer


In [275]:
words = []
# for row in cleaned_text_ls:
for row in clean_text:
  # for i in row.split():
  for i in row:
    # print(i)
    words.append(i)

In [276]:
from collections import Counter 

print('Total number of words in cleaned corpus:', len(words))
print('Number of unique words in cleaned corpus:', len(Counter(words)), len(Counter(words).most_common()))

Total number of words in cleaned corpus: 36437227
Number of unique words in cleaned corpus: 55989 55989


Total number of words in cleaned corpus: 36436399

Number of unique words in cleaned corpus: 55984 55984